// dummy-json/json generator.js
/*! JSON Generator - v1.2.0 - 2013/04/09
 * http://www.json-generator.com
 * Copyright (c) 2013 Vazha Omanashvili http://www.rainum.ru
 */
var ZeroClipboard = {
  version: "1.0.7",
  clients: {},
  moviePath: "ZeroClipboard.swf",
  nextId: 1,
  $: function (e) {
    return "string" == typeof e && (e = document.getElementById(e)), e.addClass || (e.hide = function () {
        this.style.display = "none"
      }, e.show = function () {
        this.style.display = ""
      }, e.addClass = function (e) {
        this.removeClass(e), this.className += " " + e
      }, e.removeClass = function (e) {
        for (var t = this.className.split(/\s+/), n = -1, r = 0; t.length > r; r++) t[r] == e && (n = r, r = t.length);
        return n > -1 && (t.splice(n, 1), this.className = t.join(" ")), this
      }, e.hasClass = function (e) {
        return !!this.className.match(RegExp("\\s*" + e + "\\s*"))
      }), e
  },
  setMoviePath: function (e) {
    this.moviePath = e
  },
  dispatch: function (e, t, n) {
    var r = this.clients[e];
    r && r.receiveEvent(t, n)
  },
  register: function (e, t) {
    this.clients[e] = t
  },
  getDOMObjectPosition: function (e, t) {
    for (var n = {
        left: 0,
        top: 0,
        width: e.width ? e.width : e.offsetWidth,
        height: e.height ? e.height : e.offsetHeight
      }; e && e != t;) n.left += e.offsetLeft, n.top += e.offsetTop, e = e.offsetParent;
    return n
  },
  Client: function (e) {
    this.handlers = {}, this.id = ZeroClipboard.nextId++, this.movieId = "ZeroClipboardMovie_" + this.id, ZeroClipboard.register(this.id, this), e && this.glue(e)
  }
};
ZeroClipboard.Client.prototype = {
  id: 0,
  ready: !1,
  movie: null,
  clipText: "",
  handCursorEnabled: !0,
  cssEffects: !0,
  handlers: null,
  glue: function (e, t, n) {
    this.domElement = ZeroClipboard.$(e);
    var r = 99;
    this.domElement.style.zIndex && (r = parseInt(this.domElement.style.zIndex, 10) + 1), "string" == typeof t ? t = ZeroClipboard.$(t) : t === void 0 && (t = document.getElementsByTagName("body")[0]);
    var i = ZeroClipboard.getDOMObjectPosition(this.domElement, t);
    this.div = document.createElement("div");
    var o = this.div.style;
    if (o.position = "absolute", o.left = "" + i.left + "px", o.top = "" + i.top + "px", o.width = "" + i.width + "px", o.height = "" + i.height + "px", o.zIndex = r, "object" == typeof n) for (addedStyle in n) o[addedStyle] = n[addedStyle];
    t.appendChild(this.div), this.div.innerHTML = this.getHTML(i.width, i.height)
  },
  getHTML: function (e, t) {
    var n = "",
      r = "id=" + this.id + "&width=" + e + "&height=" + t;
    if (navigator.userAgent.match(/MSIE/)) {
      var i = location.href.match(/^https/i) ? "https://" : "http://";
      n += '<object classid="clsid:d27cdb6e-ae6d-11cf-96b8-444553540000" codebase="' + i + 'download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=9,0,0,0" width="' + e + '" height="' + t + '" id="' + this.movieId + '" align="middle"><param name="allowScriptAccess" value="always" /><param name="allowFullScreen" value="false" /><param name="movie" value="' + ZeroClipboard.moviePath + '" /><param name="loop" value="false" /><param name="menu" value="false" /><param name="quality" value="best" /><param name="bgcolor" value="#ffffff" /><param name="flashvars" value="' + r + '"/><param name="wmode" value="transparent"/></object>'
    } else n += '<embed id="' + this.movieId + '" src="' + ZeroClipboard.moviePath + '" loop="false" menu="false" quality="best" bgcolor="#ffffff" width="' + e + '" height="' + t + '" name="' + this.movieId + '" align="middle" allowScriptAccess="always" allowFullScreen="false" type="application/x-shockwave-flash" pluginspage="http://www.macromedia.com/go/getflashplayer" flashvars="' + r + '" wmode="transparent" />';
    return n
  },
  hide: function () {
    this.div && (this.div.style.left = "-2000px")
  },
  show: function () {
    this.reposition()
  },
  destroy: function () {
    if (this.domElement && this.div) {
      this.hide(), this.div.innerHTML = "";
      var e = document.getElementsByTagName("body")[0];
      try {
        e.removeChild(this.div)
      } catch (t) {}
      this.domElement = null, this.div = null
    }
  },
  reposition: function (e) {
    if (e && (this.domElement = ZeroClipboard.$(e), this.domElement || this.hide()), this.domElement && this.div) {
      var t = ZeroClipboard.getDOMObjectPosition(this.domElement),
        n = this.div.style;
      n.left = "" + t.left + "px", n.top = "" + t.top + "px"
    }
  },
  setText: function (e) {
    this.clipText = e, this.ready && this.movie.setText(e)
  },
  addEventListener: function (e, t) {
    e = ("" + e).toLowerCase().replace(/^on/, ""), this.handlers[e] || (this.handlers[e] = []), this.handlers[e].push(t)
  },
  setHandCursor: function (e) {
    this.handCursorEnabled = e, this.ready && this.movie.setHandCursor(e)
  },
  setCSSEffects: function (e) {
    this.cssEffects = !! e
  },
  receiveEvent: function (e, t) {
    switch (e = ("" + e).toLowerCase().replace(/^on/, "")) {
    case "load":
      if (this.movie = document.getElementById(this.movieId), !this.movie) {
        var n = this;
        return setTimeout(function () {
            n.receiveEvent("load", null)
          }, 1), void 0
      }
      if (!this.ready && navigator.userAgent.match(/Firefox/) && navigator.userAgent.match(/Windows/)) {
        var n = this;
        return setTimeout(function () {
            n.receiveEvent("load", null)
          }, 100), this.ready = !0, void 0
      }
      this.ready = !0, this.movie.setText(this.clipText), this.movie.setHandCursor(this.handCursorEnabled);
      break;
    case "mouseover":
      this.domElement && this.cssEffects && (this.domElement.addClass("hover"), this.recoverActive && this.domElement.addClass("active"));
      break;
    case "mouseout":
      this.domElement && this.cssEffects && (this.recoverActive = !1, this.domElement.hasClass("active") && (this.domElement.removeClass("active"), this.recoverActive = !0), this.domElement.removeClass("hover"));
      break;
    case "mousedown":
      this.domElement && this.cssEffects && this.domElement.addClass("active");
      break;
    case "mouseup":
      this.domElement && this.cssEffects && (this.domElement.removeClass("active"), this.recoverActive = !1)
    }
    if (this.handlers[e]) for (var r = 0, i = this.handlers[e].length; i > r; r++) {
        var o = this.handlers[e][r];
        "function" == typeof o ? o(this, t) : "object" == typeof o && 2 == o.length ? o[0][o[1]](this, t) : "string" == typeof o && window[o](this, t)
    }
  }
}, window.CodeMirror = function () {
  "use strict";

  function e(a, u) {
    function c(e) {
      nr.onDragEvent && nr.onDragEvent(oi, T(e)) || W(e)
    }

    function m(e) {
      return e >= 0 && Tr.size > e
    }

    function g(e) {
      return C(Tr, e)
    }

    function M(e, t) {
      Pr = !0;
      for (var n = t - e.height, r = e; r; r = r.parent) r.height += n
    }

    function E(e, t) {
      return e.styles || e.highlight(wr, e.stateAfter = Jn(b(e)), nr.tabSize), e.getContent(nr.tabSize, t, nr.lineWrapping)
    }

    function I(e) {
      var t = {
        line: 0,
        ch: 0
      };
      Nt(t, {
          line: Tr.size - 1,
          ch: g(Tr.size - 1).text.length
        }, wt(e), t, t), Dr = !0
    }

    function $(e) {
      var t = [];
      return Tr.iter(0, Tr.size, function (e) {
          t.push(e.text)
        }), t.join(e || "\n")
    }

    function _() {
      sr.scrollTop != Fr && (Fr = xr.scrollTop = sr.scrollTop, en([]))
    }

    function et() {
      nr.fixedGutter && gr.style.left != xr.scrollLeft + "px" && (gr.style.left = xr.scrollLeft + "px"), xr.scrollTop != Fr && (Fr = xr.scrollTop, sr.scrollTop != Fr && (sr.scrollTop = Fr), en([])), nr.onScroll && nr.onScroll(oi)
    }

    function rt(e) {
      function t(t) {
        ft && (xr.draggable = !1), Nr = !1, h(), d(), 10 > Math.abs(e.clientX - t.clientX) + Math.abs(e.clientY - t.clientY) && (A(t), cn(a.line, a.ch, !0), Zt())
      }

      function n(e) {
        if ("single" == s) ln(a, e);
        else if ("double" == s) {
          var t = vn(e);
          V(e, p) ? ln(t.from, m) : ln(p, t.to)
        } else "triple" == s && (V(e, p) ? ln(m, hn({
                  line: e.line,
                  ch: 0
                })) : ln(p, hn({
                  line: e.line + 1,
                  ch: 0
                })))
      }

      function r(e) {
        var t = Kn(e, !0);
        if (t && !F(t, f)) {
          Sr || Wt(), f = t, n(t), Dr = !1;
          var i = _t();
          (t.line >= i.to || t.line < i.from) && (c = setTimeout(er(function () {
                  r(e)
                }), 150))
        }
      }

      function i(e) {
        clearTimeout(c);
        var t = Kn(e);
        t && n(t), A(e), Zt(), Dr = !0, g(), h()
      }
      an(D(e, "shiftKey"));
      for (var o = L(e); o != Cr; o = o.parentNode) if (o.parentNode == yr && o != vr) return;
      for (var o = L(e); o != Cr; o = o.parentNode) if (o.parentNode == mr) return nr.onGutterClick && nr.onGutterClick(oi, J(mr.childNodes, o) + Yr, e), A(e);
      var a = Kn(e);
      switch (N(e)) {
      case 3:
        return at && Un(e), void 0;
      case 2:
        return a && cn(a.line, a.ch, !0), setTimeout(Zt, 20), A(e), void 0
      }
      if (!a) return L(e) == xr && A(e), void 0;
      Sr || Wt();
      var l = +new Date,
        s = "single";
      if (Lr && Lr.time > l - 400 && F(Lr.pos, a)) s = "triple", A(e), setTimeout(Zt, 20), yn(a.line);
      else if (Wr && Wr.time > l - 400 && F(Wr.pos, a)) {
        s = "double", Lr = {
          time: l,
          pos: a
        }, A(e);
        var u = vn(a);
        ln(u.from, u.to)
      } else Wr = {
          time: l,
          pos: a
      };
      var c, f = a;
      if (nr.dragDrop && xt && !nr.readOnly && !F(jr.from, jr.to) && !V(a, jr.from) && !V(jr.to, a) && "single" == s) {
        ft && (xr.draggable = !0);
        var h = z(document, "mouseup", er(t), !0),
          d = z(xr, "drop", er(t), !0);
        return Nr = !0, xr.dragDrop && xr.dragDrop(), void 0
      }
      A(e), "single" == s && cn(a.line, a.ch, !0);
      var p = jr.from,
        m = jr.to,
        g = z(document, "mousemove", er(function (e) {
              clearTimeout(c), A(e), lt || N(e) ? r(e) : i(e)
            }), !0),
        h = z(document, "mouseup", er(i), !0)
    }

    function ht(e) {
      for (var t = L(e); t != Cr; t = t.parentNode) if (t.parentNode == mr) return A(e);
      A(e)
    }

    function pt(e) {
      if (!nr.onDragEvent || !nr.onDragEvent(oi, T(e))) {
        A(e);
        var t = Kn(e, !0),
          n = e.dataTransfer.files;
        if (t && !nr.readOnly) if (n && n.length && window.FileReader && window.File) for (var r = n.length, i = Array(r), o = 0, a = function (e, n) {
                  var a = new FileReader;
                  a.onload = function () {
                    i[n] = a.result, ++o == r && (t = hn(t), er(function () {
                          var e = Rt(i.join(""), t, t);
                          ln(t, e)
                        })())
                  }, a.readAsText(e)
                }, l = 0; r > l; ++l) a(n[l], l);
          else {
            if (Nr && !V(t, jr.from) && !V(jr.to, t)) return;
            try {
              var i = e.dataTransfer.getData("Text");
              i && tr(function () {
                  var e = jr.from,
                    n = jr.to;
                  ln(t, t), Nr && Rt("", e, n), jt(i), Zt()
                })
            } catch (e) {}
          }
      }
    }

    function vt(e) {
      var t = Gt();
      e.dataTransfer.setData("Text", t), e.dataTransfer.setDragImage && e.dataTransfer.setDragImage(K("img"), 0, 0)
    }

    function Ct(e, t) {
      if ("string" == typeof e && (e = it[e], !e)) return !1;
      var n = Er;
      try {
        nr.readOnly && (Gr = !0), t && (Er = null), e(oi)
      } catch (r) {
        if (r != yt) throw r;
        return !1
      } finally {
        Er = n, Gr = !1
      }
      return !0
    }

    function bt(e) {
      function i() {
        c = !0
      }
      var o = t(nr.keyMap),
        a = o.auto;
      clearTimeout(ii), a && !r(e) && (ii = setTimeout(function () {
            t(nr.keyMap) == o && (nr.keyMap = a.call ? a.call(null, oi) : a)
          }, 50));
      var l = kt[D(e, "keyCode")],
        s = !1,
        u = dt && X;
      if (null == l || e.altGraphKey) return !1;
      D(e, "altKey") && (l = "Alt-" + l), D(e, u ? "metaKey" : "ctrlKey") && (l = "Ctrl-" + l), D(e, u ? "ctrlKey" : "metaKey") && (l = "Cmd-" + l);
      var c = !1;
      return s = D(e, "shiftKey") ? n("Shift-" + l, nr.extraKeys, nr.keyMap, function (e) {
          return Ct(e, !0)
        }, i) || n(l, nr.extraKeys, nr.keyMap, function (e) {
          return "string" == typeof e && /^go[A-Z]/.test(e) ? Ct(e) : void 0
        }, i) : n(l, nr.extraKeys, nr.keyMap, Ct, i), c && (s = !1), s && (A(e), Yn(), lt && (e.oldKeyCode = e.keyCode, e.keyCode = 0)), s
    }

    function Mt(e, t) {
      var r = n("'" + t + "'", nr.extraKeys, nr.keyMap, function (e) {
          return Ct(e, !0)
        });
      return r && (A(e), Yn()), r
    }

    function Tt(e) {
      if (Sr || Wt(), lt && 27 == e.keyCode && (e.returnValue = !1), Xr && Yt() && (Xr = !1), !nr.onKeyEvent || !nr.onKeyEvent(oi, T(e))) {
        var t = D(e, "keyCode");
        an(16 == t || D(e, "shiftKey"));
        var n = bt(e);
        dt && (ai = n ? t : null, !n && 88 == t && D(e, X ? "metaKey" : "ctrlKey") && jt(""))
      }
    }

    function At(e) {
      if (Xr && Yt(), !nr.onKeyEvent || !nr.onKeyEvent(oi, T(e))) {
        var t = D(e, "keyCode"),
          n = D(e, "charCode");
        if (dt && t == ai) return ai = null, A(e), void 0;
        if (!(dt && (!e.which || 10 > e.which) || mt) || !bt(e)) {
          var r = String.fromCharCode(null == n ? t : n);
          nr.electricChars && wr.electricChars && nr.smartIndent && !nr.readOnly && wr.electricChars.indexOf(r) > -1 && setTimeout(er(function () {
                Cn(jr.to.line, "smart")
              }), 75), Mt(e, r) || Ut()
        }
      }
    }

    function Et(e) {
      nr.onKeyEvent && nr.onKeyEvent(oi, T(e)) || 16 == D(e, "keyCode") && (Er = null)
    }

    function Wt() {
      "nocursor" != nr.readOnly && (Sr || (nr.onFocus && nr.onFocus(oi), Sr = !0, -1 == xr.className.search(/\bCodeMirror-focused\b/) && (xr.className += " CodeMirror-focused")), Kt(), Yn())
    }

    function Lt() {
      Sr && (nr.onBlur && nr.onBlur(oi), Sr = !1, Rr && er(function () {
            Rr && (Rr(), Rr = null)
          })(), xr.className = xr.className.replace(" CodeMirror-focused", "")), clearInterval(br), setTimeout(function () {
          Sr || (Er = null)
        }, 150)
    }

    function Nt(e, t, n, r, i) {
      if (!Gr) {
        var o = [];
        if (Tr.iter(e.line, t.line + 1, function (e) {
              o.push(p(e.text, e.markedSpans))
            }), ei) for (ei.addChange(e.line, n.length, o); ei.done.length > nr.undoDepth;) ei.done.shift();
        var a = f(d(o[0]), d(R(o)), e.ch, t.ch, n);
        It(e, t, a, r, i)
      }
    }

    function Dt(e, t) {
      if (e.length) {
        for (var n = e.pop(), r = [], i = n.length - 1; i >= 0; i -= 1) {
          var o = n[i],
            a = [],
            l = o.start + o.added;
          Tr.iter(o.start, l, function (e) {
              a.push(p(e.text, e.markedSpans))
            }), r.push({
              start: o.start,
              added: o.old.length,
              old: a
            });
          var s = {
            line: o.start + o.old.length - 1,
            ch: Z(h(R(a)), h(R(o.old)))
          };
          It({
              line: o.start,
              ch: 0
            }, {
              line: l - 1,
              ch: g(l - 1).text.length
            }, o.old, s, s)
        }
        Dr = !0, t.push(r)
      }
    }

    function zt() {
      Dt(ei.done, ei.undone)
    }

    function Ot() {
      Dt(ei.undone, ei.done)
    }

    function It(e, t, n, r, i) {
      function o(e) {
        return Math.min(t.line, t.line + C) >= e ? e : e + C
      }
      if (!Gr) {
        var a = !1,
          l = Jr.text.length;
        nr.lineWrapping || Tr.iter(e.line, t.line + 1, function (e) {
            return e.hidden || e.text.length != l ? void 0 : (a = !0, !0)
          }), (e.line != t.line || n.length > 1) && (Pr = !0);
        var s = t.line - e.line,
          u = g(e.line),
          c = g(t.line),
          f = R(n);
        if (0 == e.ch && 0 == t.ch && "" == h(f)) {
          for (var p = [], m = 0, y = n.length - 1; y > m; ++m) p.push(new v(h(n[m]), d(n[m])));
          c.update(c.text, d(f)), s && Tr.remove(e.line, s, Br), p.length && Tr.insert(e.line, p)
        } else if (u == c) if (1 == n.length) u.update(u.text.slice(0, e.ch) + h(n[0]) + u.text.slice(t.ch), d(n[0]));
          else {
            for (var p = [], m = 1, y = n.length - 1; y > m; ++m) p.push(new v(h(n[m]), d(n[m])));
            p.push(new v(h(f) + u.text.slice(t.ch), d(f))), u.update(u.text.slice(0, e.ch) + h(n[0]), d(n[0])), Tr.insert(e.line + 1, p)
          } else if (1 == n.length) u.update(u.text.slice(0, e.ch) + h(n[0]) + c.text.slice(t.ch), d(n[0])), Tr.remove(e.line + 1, s, Br);
        else {
          var p = [];
          u.update(u.text.slice(0, e.ch) + h(n[0]), d(n[0])), c.update(h(f) + c.text.slice(t.ch), d(f));
          for (var m = 1, y = n.length - 1; y > m; ++m) p.push(new v(h(n[m]), d(n[m])));
          s > 1 && Tr.remove(e.line + 1, s - 1, Br), Tr.insert(e.line + 1, p)
        } if (nr.lineWrapping) {
          var x = Math.max(5, xr.clientWidth / Fn() - 3);
          Tr.iter(e.line, e.line + n.length, function (e) {
              if (!e.hidden) {
                var t = Math.ceil(e.text.length / x) || 1;
                t != e.height && M(e, t)
              }
            })
        } else Tr.iter(e.line, e.line + n.length, function (e) {
              var t = e.text;
              !e.hidden && t.length > l && (Jr = e, l = t.length, Qr = !0, a = !1)
            }), a && (qr = !0);
        Ar = Math.min(Ar, e.line), Qn(400);
        var C = n.length - s - 1;
        if (Or.push({
              from: e.line,
              to: t.line + 1,
              diff: C
            }), nr.onChange) {
          for (var m = 0; n.length > m; ++m) "string" != typeof n[m] && (n[m] = n[m].text);
          var b = {
            from: e,
            to: t,
            text: n
          };
          if (Ir) {
            for (var w = Ir; w.next; w = w.next);
            w.next = b
          } else Ir = b
        }
        sn(hn(r), hn(i), o(jr.from.line), o(jr.to.line))
      }
    }

    function Ht() {
      var e = Tr.height * jn() + 2 * Vn();
      return .99 * e > xr.offsetHeight ? e : !1
    }

    function Pt(e) {
      var t = Ht();
      sr.style.display = t ? "block" : "none", t ? (lr.style.height = yr.style.minHeight = t + "px", sr.style.height = xr.clientHeight + "px", null != e && (sr.scrollTop = xr.scrollTop = e, ft && setTimeout(function () {
              sr.scrollTop == e && (sr.scrollTop = e + (e ? -1 : 1), sr.scrollTop = e)
            }, 0))) : yr.style.minHeight = "", vr.style.top = Ur * jn() + "px"
    }

    function Bt() {
      Jr = g(0), Qr = !0;
      var e = Jr.text.length;
      Tr.iter(1, Tr.size, function (t) {
          var n = t.text;
          !t.hidden && n.length > e && (e = n.length, Jr = t)
        }), qr = !1
    }

    function Rt(e, t, n) {
      function r(r) {
        if (V(r, t)) return r;
        if (!V(n, r)) return i;
        var o = r.line + e.length - (n.line - t.line) - 1,
          a = r.ch;
        return r.line == n.line && (a += R(e).length - (n.ch - (n.line == t.line ? t.ch : 0))), {
          line: o,
          ch: a
        }
      }
      t = hn(t), n = n ? hn(n) : t, e = wt(e);
      var i;
      return Ft(e, t, n, function (e) {
          return i = e, {
            from: r(jr.from),
            to: r(jr.to)
          }
        }), i
    }

    function jt(e, t) {
      Ft(wt(e), jr.from, jr.to, function (e) {
          return "end" == t ? {
            from: e,
            to: e
          } : "start" == t ? {
            from: jr.from,
            to: jr.from
          } : {
            from: jr.from,
            to: e
          }
        })
    }

    function Ft(e, t, n, r) {
      var i = 1 == e.length ? e[0].length + t.ch : R(e).length,
        o = r({
            line: t.line + e.length - 1,
            ch: i
          });
      Nt(t, n, e, o.from, o.to)
    }

    function Vt(e, t, n) {
      var r = e.line,
        i = t.line;
      if (r == i) return g(r).text.slice(e.ch, t.ch);
      var o = [g(r).text.slice(e.ch)];
      return Tr.iter(r + 1, i, function (e) {
          o.push(e.text)
        }), o.push(g(i).text.slice(0, t.ch)), o.join(n || "\n")
    }

    function Gt(e) {
      return Vt(jr.from, jr.to, e)
    }

    function Kt() {
      Xr || kr.set(nr.pollInterval, function () {
          Yt(), Sr && Kt()
        })
    }

    function Ut() {
      function e() {
        var n = Yt();
        n || t ? (Xr = !1, Kt()) : (t = !0, kr.set(60, e))
      }
      var t = !1;
      Xr = !0, kr.set(20, e)
    }

    function Yt() {
      if (!Sr || St(or) || nr.readOnly) return !1;
      var e = or.value;
      if (e == li) return !1;
      mi || Xn(), Er = null;
      for (var t = 0, n = Math.min(li.length, e.length); n > t && li[t] == e[t];)++t;
      return li.length > t ? jr.from = {
        line: jr.from.line,
        ch: jr.from.ch - (li.length - t)
      } : Vr && F(jr.from, jr.to) && !Kr && (jr.to = {
          line: jr.to.line,
          ch: Math.min(g(jr.to.line).text.length, jr.to.ch + (e.length - t))
        }), jt(e.slice(t), "end"), e.length > 1e3 ? or.value = li = "" : li = e, mi || _n(), Kr = !1, !0
    }

    function $t(e) {
      F(jr.from, jr.to) ? e && (li = or.value = "") : (li = "", or.value = Gt(), Sr && j(or))
    }

    function Zt() {
      "nocursor" != nr.readOnly && or.focus()
    }

    function Jt() {
      var e = qt();
      if (Qt(e.x, e.y, e.x, e.yBot), Sr) {
        var t = yr.getBoundingClientRect(),
          n = null;
        if (0 > e.y + t.top ? n = !0 : e.y + t.top + jn() > (window.innerHeight || document.documentElement.clientHeight) && (n = !1), null != n) {
          var r = "none" == fr.style.display;
          r && (fr.style.display = "", fr.style.left = e.x + "px", fr.style.top = e.y - Ur + "px"), fr.scrollIntoView(n), r && (fr.style.display = "none")
        }
      }
    }

    function qt() {
      var e = Pn(jr.inverted ? jr.from : jr.to),
        t = nr.lineWrapping ? Math.min(e.x, pr.offsetWidth) : e.x;
      return {
        x: t,
        y: e.y,
        yBot: e.yBot
      }
    }

    function Qt(e, t, n, r) {
      var i = Xt(e, t, n, r);
      null != i.scrollLeft && (xr.scrollLeft = i.scrollLeft), null != i.scrollTop && (sr.scrollTop = xr.scrollTop = i.scrollTop)
    }

    function Xt(e, t, n, r) {
      var i = Gn(),
        o = Vn();
      t += o, r += o, e += i, n += i;
      var a = xr.clientHeight,
        l = sr.scrollTop,
        s = {}, u = Ht() || 1 / 0,
        c = o + 10 > t,
        f = r + o > u - 10;
      l > t ? s.scrollTop = c ? 0 : Math.max(0, t) : r > l + a && (s.scrollTop = (f ? u : r) - a);
      var h = xr.clientWidth,
        d = xr.scrollLeft,
        p = nr.fixedGutter ? gr.clientWidth : 0,
        m = p + i + 10 > e;
      return d + p > e || m ? (m && (e = 0), s.scrollLeft = Math.max(0, e - 10 - p)) : n > h + d - 3 && (s.scrollLeft = n + 10 - h), s
    }

    function _t(e) {
      var t = jn(),
        n = (null != e ? e : sr.scrollTop) - Vn(),
        r = Math.max(0, Math.floor(n / t)),
        i = Math.ceil((n + xr.clientHeight) / t);
      return {
        from: w(Tr, r),
        to: w(Tr, i)
      }
    }

    function en(e, t, n) {
      function r() {
        var e = ur.firstChild,
          t = !1;
        return Tr.iter(Yr, $r, function (n) {
            if (e) {
              if (!n.hidden) {
                var r = Math.round(e.offsetHeight / f) || 1;
                n.height != r && (M(n, r), Pr = t = !0)
              }
              e = e.nextSibling
            }
          }), t
      }
      if (!xr.clientWidth) return Yr = $r = Ur = 0, void 0;
      var i = _t(n);
      if (e !== !0 && 0 == e.length && i.from > Yr && $r > i.to) return Pt(n), void 0;
      var o = Math.max(i.from - 100, 0),
        a = Math.min(Tr.size, i.to + 100);
      o > Yr && 20 > o - Yr && (o = Yr), $r > a && 20 > $r - a && (a = Math.min(Tr.size, $r));
      for (var l = e === !0 ? [] : tn([{
              from: Yr,
              to: $r,
              domStart: 0
            }
          ], e), s = 0, u = 0; l.length > u; ++u) {
        var c = l[u];
        o > c.from && (c.domStart += o - c.from, c.from = o), c.to > a && (c.to = a), c.from >= c.to ? l.splice(u--, 1) : s += c.to - c.from
      }
      if (s == a - o && o == Yr && a == $r) return Pt(n), void 0;
      l.sort(function (e, t) {
          return e.domStart - t.domStart
        });
      var f = jn(),
        h = gr.style.display;
      ur.style.display = "none", nn(o, a, l), ur.style.display = gr.style.display = "";
      var d = o != Yr || a != $r || Zr != xr.clientHeight + f;
      if (d && (Zr = xr.clientHeight + f), (o != Yr || a != $r && nr.onViewportChange) && setTimeout(function () {
            nr.onViewportChange && nr.onViewportChange(oi, o, a)
          }), Yr = o, $r = a, Ur = S(Tr, o), Qn(100), ur.childNodes.length != $r - Yr) throw Error("BAD PATCH! " + JSON.stringify(l) + " size=" + ($r - Yr) + " nodes=" + ur.childNodes.length);
      return nr.lineWrapping && r(), gr.style.display = h, (d || Pr) && rn() && nr.lineWrapping && r() && rn(), Pt(n), on(), !t && nr.onUpdate && nr.onUpdate(oi), !0
    }

    function tn(e, t) {
      for (var n = 0, r = t.length || 0; r > n; ++n) {
        for (var i = t[n], o = [], a = i.diff || 0, l = 0, s = e.length; s > l; ++l) {
          var u = e[l];
          i.to <= u.from && i.diff ? o.push({
              from: u.from + a,
              to: u.to + a,
              domStart: u.domStart
            }) : i.to <= u.from || i.from >= u.to ? o.push(u) : (i.from > u.from && o.push({
                from: u.from,
                to: i.from,
                domStart: u.domStart
              }), i.to < u.to && o.push({
                from: i.to + a,
                to: u.to + a,
                domStart: u.domStart + (i.to - u.from)
              }))
        }
        e = o
      }
      return e
    }

    function nn(e, t, n) {
      function r(e) {
        var t = e.nextSibling;
        return e.parentNode.removeChild(e), t
      }
      if (n.length) {
        for (var i = 0, o = ur.firstChild, a = 0; n.length > a; ++a) {
          for (var l = n[a]; l.domStart > i;) o = r(o), i++;
          for (var s = 0, u = l.to - l.from; u > s; ++s) o = o.nextSibling, i++
        }
        for (; o;) o = r(o)
      } else U(ur);
      var c = n.shift(),
        o = ur.firstChild,
        s = e;
      Tr.iter(e, t, function (e) {
          if (c && c.to == s && (c = n.shift()), !c || c.from > s) {
            if (e.hidden) var t = K("pre");
            else {
              var t = E(e);
              if (e.className && (t.className = e.className), e.bgClassName) {
                var r = K("pre", " ", e.bgClassName, "position: absolute; left: 0; right: 0; top: 0; bottom: 0; z-index: -2");
                t = K("div", [r, t], null, "position: relative")
              }
            }
            ur.insertBefore(t, o)
          } else o = o.nextSibling;
          ++s
        })
    }

    function rn() {
      if (nr.gutter || nr.lineNumbers) {
        var e = vr.offsetHeight,
          t = xr.clientHeight;
        gr.style.height = (2 > e - t ? t : e) + "px";
        var n, r = document.createDocumentFragment(),
          i = Yr;
        if (Tr.iter(Yr, Math.max($r, Yr + 1), function (e) {
              if (e.hidden) r.appendChild(K("pre"));
              else {
                var t = e.gutterMarker,
                  o = nr.lineNumbers ? nr.lineNumberFormatter(i + nr.firstLineNumber) : null;
                t && t.text ? o = t.text.replace("%N%", null != o ? o : "") : null == o && (o = " ");
                var a = r.appendChild(K("pre", null, t && t.style));
                a.innerHTML = o;
                for (var l = 1; e.height > l; ++l) a.appendChild(K("br")), a.appendChild(document.createTextNode(" "));
                t || (n = i)
              }++i
            }), gr.style.display = "none", Y(mr, r), null != n && nr.lineNumbers) {
          for (var o = mr.childNodes[n - Yr], a = (Tr.size + "").length, l = P(o.firstChild), s = ""; a > l.length + s.length;) s += " ";
          s && o.insertBefore(document.createTextNode(s), o.firstChild)
        }
        gr.style.display = "";
        var u = Math.abs((parseInt(pr.style.marginLeft) || 0) - gr.offsetWidth) > 2;
        return pr.style.marginLeft = gr.offsetWidth + "px", Pr = !1, u
      }
    }

    function on() {
      var e = F(jr.from, jr.to),
        t = Pn(jr.from, !0),
        n = e ? t : Pn(jr.to, !0),
        r = jr.inverted ? t : n,
        i = jn(),
        o = H(Cr),
        a = H(ur);
      if (ar.style.top = Math.max(0, Math.min(xr.offsetHeight, r.y + a.top - o.top)) + "px", ar.style.left = Math.max(0, Math.min(xr.offsetWidth, r.x + a.left - o.left)) + "px", e) fr.style.top = r.y + "px", fr.style.left = (nr.lineWrapping ? Math.min(r.x, pr.offsetWidth) : r.x) + "px", fr.style.display = "", cr.style.display = "none";
      else {
        var l = t.y == n.y,
          s = document.createDocumentFragment(),
          u = pr.clientWidth || pr.offsetWidth,
          c = pr.clientHeight || pr.offsetHeight,
          f = function (e, t, n, r) {
            var i = ct ? "width: " + (n ? u - n - e : u) + "px" : "right: " + n + "px";
            s.appendChild(K("div", null, "CodeMirror-selected", "position: absolute; left: " + e + "px; top: " + t + "px; " + i + "; height: " + r + "px"))
          };
        if (jr.from.ch && t.y >= 0) {
          var h = l ? u - n.x : 0;
          f(t.x, t.y, h, i)
        }
        var d = Math.max(0, t.y + (jr.from.ch ? i : 0)),
          p = Math.min(n.y, c) - d;
        p > .2 * i && f(0, d, 0, p), l && jr.from.ch || !(c - .5 * i > n.y) || f(0, n.y, u - n.x, i), Y(cr, s), fr.style.display = "none", cr.style.display = ""
      }
    }

    function an(e) {
      Er = e ? Er || (jr.inverted ? jr.to : jr.from) : null
    }

    function ln(e, t) {
      var n = Er && hn(Er);
      n && (V(n, e) ? e = n : V(t, n) && (t = n)), sn(e, t), zr = !0
    }

    function sn(e, t, n, r) {
      if (_r = null, null == n && (n = jr.from.line, r = jr.to.line), !F(jr.from, e) || !F(jr.to, t)) {
        if (V(t, e)) {
          var i = t;
          t = e, e = i
        }
        if (e.line != n) {
          var o = un(e, n, jr.from.ch);
          o ? e = o : On(e.line, !1)
        }
        if (t.line != r && (t = un(t, r, jr.to.ch)), F(e, t) ? jr.inverted = !1 : F(e, jr.to) ? jr.inverted = !1 : F(t, jr.from) && (jr.inverted = !0), nr.autoClearEmptyLines && F(jr.from, jr.to)) {
          var a = jr.inverted ? e : t;
          if (a.line != jr.from.line && jr.from.line < Tr.size) {
            var l = g(jr.from.line);
            /^\s+$/.test(l.text) && setTimeout(er(function () {
                  if (l.parent && /^\s+$/.test(l.text)) {
                    var e = b(l);
                    Rt("", {
                        line: e,
                        ch: 0
                      }, {
                        line: e,
                        ch: l.text.length
                      })
                  }
                }, 10))
          }
        }
        jr.from = e, jr.to = t, Hr = !0
      }
    }

    function un(e, t, n) {
      function r(t) {
        for (var r = e.line + t, i = 1 == t ? Tr.size : -1; r != i;) {
          var a = g(r);
          if (!a.hidden) {
            var l = e.ch;
            return (o || l > n || l > a.text.length) && (l = a.text.length), {
              line: r,
              ch: l
            }
          }
          r += t
        }
      }
      var i = g(e.line),
        o = e.ch == i.text.length && e.ch != n;
      return i.hidden ? e.line >= t ? r(1) || r(-1) : r(-1) || r(1) : e
    }

    function cn(e, t, n) {
      var r = hn({
          line: e,
          ch: t || 0
        });
      (n ? ln : sn)(r, r)
    }

    function fn(e) {
      return Math.max(0, Math.min(e, Tr.size - 1))
    }

    function hn(e) {
      if (0 > e.line) return {
          line: 0,
          ch: 0
      };
      if (e.line >= Tr.size) return {
          line: Tr.size - 1,
          ch: g(Tr.size - 1).text.length
      };
      var t = e.ch,
        n = g(e.line).text.length;
      return null == t || t > n ? {
        line: e.line,
        ch: n
      } : 0 > t ? {
        line: e.line,
        ch: 0
      } : e
    }

    function dn(e, t) {
      function n() {
        for (var t = o + e, n = 0 > e ? -1 : Tr.size; t != n; t += e) {
          var r = g(t);
          if (!r.hidden) return o = t, l = r, !0
        }
      }

      function r(t) {
        if (a == (0 > e ? 0 : l.text.length)) {
          if (t || !n()) return !1;
          a = 0 > e ? l.text.length : 0
        } else a += e;
        return !0
      }
      var i = jr.inverted ? jr.from : jr.to,
        o = i.line,
        a = i.ch,
        l = g(o);
      if ("char" == t) r();
      else if ("column" == t) r(!0);
      else if ("word" == t) for (var s = !1; !(0 > e) || r();) {
          if (q(l.text.charAt(a))) s = !0;
          else if (s) {
            0 > e && (e = 1, r());
            break
          }
          if (e > 0 && !r()) break
      }
      return {
        line: o,
        ch: a
      }
    }

    function pn(e, t) {
      var n = 0 > e ? jr.from : jr.to;
      (Er || F(jr.from, jr.to)) && (n = dn(e, t)), cn(n.line, n.ch, !0)
    }

    function mn(e, t) {
      F(jr.from, jr.to) ? 0 > e ? Rt("", dn(e, t), jr.to) : Rt("", jr.from, dn(e, t)) : Rt("", jr.from, jr.to), zr = !0
    }

    function gn(e, t) {
      var n = Pn(jr.inverted ? jr.from : jr.to, !0);
      if (null != _r && (n.x = _r), "page" == t) var r = Math.min(xr.clientHeight, window.innerHeight || document.documentElement.clientHeight),
      i = Bn(n.x, n.y + r * e);
      else if ("line" == t) var o = jn(),
      i = Bn(n.x, n.y + .5 * o + e * o);
      "page" == t && (sr.scrollTop += Pn(i, !0).y - n.y), cn(i.line, i.ch, !0), _r = n.x
    }

    function vn(e) {
      var t = g(e.line).text,
        n = e.ch,
        r = e.ch;
      if (t) {
        e.after === !1 || r == t.length ? --n : ++r;
        for (var i = t.charAt(n), o = q(i) ? q : /\s/.test(i) ? function (e) {
              return /\s/.test(e)
            } : function (e) {
              return !/\s/.test(e) && q(e)
            }; n > 0 && o(t.charAt(n - 1));)--n;
        for (; t.length > r && o(t.charAt(r));)++r
      }
      return {
        from: {
          line: e.line,
          ch: n
        },
        to: {
          line: e.line,
          ch: r
        }
      }
    }

    function yn(e) {
      ln({
          line: e,
          ch: 0
        }, hn({
            line: e + 1,
            ch: 0
          }))
    }

    function xn(e) {
      if (F(jr.from, jr.to)) return Cn(jr.from.line, e);
      for (var t = jr.to.line - (jr.to.ch ? 0 : 1), n = jr.from.line; t >= n; ++n) Cn(n, e)
    }

    function Cn(e, t) {
      if (t || (t = "add"), "smart" == t) if (wr.indent) var n = Jn(e);
        else t = "prev";
      var r, i = g(e),
        o = i.indentation(nr.tabSize),
        a = i.text.match(/^\s*/)[0];
      "smart" == t && (r = wr.indent(n, i.text.slice(a.length), i.text), r == yt && (t = "prev")), "prev" == t ? r = e ? g(e - 1).indentation(nr.tabSize) : 0 : "add" == t ? r = o + nr.indentUnit : "subtract" == t && (r = o - nr.indentUnit), r = Math.max(0, r);
      var l = "",
        s = 0;
      if (nr.indentWithTabs) for (var u = Math.floor(r / nr.tabSize); u; --u) s += nr.tabSize, l += " ";
      r > s && (l += B(r - s)), l != a && Rt(l, {
          line: e,
          ch: 0
        }, {
          line: e,
          ch: a.length
        }), i.stateAfter = null
    }

    function bn() {
      wr = e.getMode(nr, nr.mode), Tr.iter(0, Tr.size, function (e) {
          e.stateAfter = null
        }), Ar = 0, Qn(100)
    }

    function wn() {
      var e = nr.gutter || nr.lineNumbers;
      gr.style.display = e ? "" : "none", e ? Pr = !0 : ur.parentNode.style.marginLeft = 0
    }

    function Sn() {
      if (nr.lineWrapping) {
        Cr.className += " CodeMirror-wrap";
        var e = xr.clientWidth / Fn() - 3;
        Tr.iter(0, Tr.size, function (t) {
            if (!t.hidden) {
              var n = Math.ceil(t.text.length / e) || 1;
              1 != n && M(t, n)
            }
          }), pr.style.minWidth = hr.style.left = ""
      } else Cr.className = Cr.className.replace(" CodeMirror-wrap", ""), Bt(), Tr.iter(0, Tr.size, function (e) {
            1 == e.height || e.hidden || M(e, 1)
          });
      Or.push({
          from: 0,
          to: Tr.size
        })
    }

    function kn() {
      xr.className = xr.className.replace(/\s*cm-s-\S+/g, "") + nr.theme.replace(/(^|\s)\s*/g, " cm-s-")
    }

    function Mn() {
      var e = ot[nr.keyMap].style;
      Cr.className = Cr.className.replace(/\s*cm-keymap-\S+/g, "") + (e ? " cm-keymap-" + e : "")
    }

    function Tn(e, t) {
      this.lines = [], this.type = e, t && (this.style = t)
    }

    function An(e, t, n, r) {
      e = hn(e), t = hn(t);
      var i = new Tn("range", n);
      if (r) for (var o in r) r.hasOwnProperty(o) && (i[o] = r[o]);
      var a = e.line;
      return Tr.iter(a, t.line + 1, function (n) {
          var r = {
            from: a == e.line ? e.ch : null,
            to: a == t.line ? t.ch : null,
            marker: i
          };
          n.markedSpans = (n.markedSpans || []).concat([r]), i.lines.push(n), ++a
        }), Or.push({
          from: e.line,
          to: t.line + 1
        }), i
    }

    function En(e) {
      e = hn(e);
      var t = new Tn("bookmark"),
        n = g(e.line);
      ei.addChange(e.line, 1, [p(n.text, n.markedSpans)], !0);
      var r = {
        from: e.ch,
        to: e.ch,
        marker: t
      };
      return n.markedSpans = (n.markedSpans || []).concat([r]), t.lines.push(n), t
    }

    function Wn(e) {
      e = hn(e);
      var t = [],
        n = g(e.line).markedSpans;
      if (n) for (var r = 0; n.length > r; ++r) {
          var i = n[r];
          (null == i.from || i.from <= e.ch) && (null == i.to || i.to >= e.ch) && t.push(i.marker)
      }
      return t
    }

    function Ln(e, t, n) {
      return "number" == typeof e && (e = g(fn(e))), e.gutterMarker = {
        text: t,
        style: n
      }, Pr = !0, e
    }

    function Nn(e) {
      "number" == typeof e && (e = g(fn(e))), e.gutterMarker = null, Pr = !0
    }

    function Dn(e, t) {
      var n = e,
        r = e;
      return "number" == typeof e ? r = g(fn(e)) : n = b(e), null == n ? null : t(r, n) ? (Or.push({
            from: n,
            to: n + 1
          }), r) : null
    }

    function zn(e, t, n) {
      return Dn(e, function (e) {
          return e.className != t || e.bgClassName != n ? (e.className = t, e.bgClassName = n, !0) : void 0
        })
    }

    function On(e, t) {
      return Dn(e, function (e, n) {
          if (e.hidden != t) {
            e.hidden = t, nr.lineWrapping || (t && e.text.length == Jr.text.length ? qr = !0 : !t && e.text.length > Jr.text.length && (Jr = e, qr = !1)), M(e, t ? 0 : 1);
            var r = jr.from.line,
              i = jr.to.line;
            if (t && (r == n || i == n)) {
              var o = r == n ? un({
                  line: r,
                  ch: 0
                }, r, 0) : jr.from,
                a = i == n ? un({
                    line: i,
                    ch: 0
                  }, i, 0) : jr.to;
              if (!a) return;
              sn(o, a)
            }
            return Pr = !0
          }
        })
    }

    function In(e) {
      if ("number" == typeof e) {
        if (!m(e)) return null;
        var t = e;
        if (e = g(e), !e) return null
      } else {
        var t = b(e);
        if (null == t) return null
      }
      var n = e.gutterMarker;
      return {
        line: t,
        handle: e,
        text: e.text,
        markerText: n && n.text,
        markerClass: n && n.style,
        lineClass: e.className,
        bgClass: e.bgClassName
      }
    }

    function Hn(e, t) {
      if (0 == t) return {
          top: 0,
          left: 0
      };
      var n = E(e, t);
      Y(dr, n);
      var r = n.anchor,
        i = r.offsetTop,
        o = r.offsetLeft;
      if (lt && 0 == i && 0 == o) {
        var a = K("span", "x");
        r.parentNode.insertBefore(a, r.nextSibling), i = a.offsetTop
      }
      return {
        top: i,
        left: o
      }
    }

    function Pn(e, t) {
      var n, r = jn(),
        i = r * (S(Tr, e.line) - (t ? Ur : 0));
      if (0 == e.ch) n = 0;
      else {
        var o = Hn(g(e.line), e.ch);
        n = o.left, nr.lineWrapping && (i += Math.max(0, o.top))
      }
      return {
        x: n,
        y: i,
        yBot: i + r
      }
    }

    function Bn(e, t) {
      function n(e) {
        var t = Hn(l, e);
        if (u) {
          var n = Math.round(t.top / r);
          return h = n != c, Math.max(0, t.left + (n - c) * xr.clientWidth)
        }
        return t.left
      }
      var r = jn(),
        i = Fn(),
        o = Ur + Math.floor(t / r);
      if (0 > o) return {
          line: 0,
          ch: 0
      };
      var a = w(Tr, o);
      if (a >= Tr.size) return {
          line: Tr.size - 1,
          ch: g(Tr.size - 1).text.length
      };
      var l = g(a),
        s = l.text,
        u = nr.lineWrapping,
        c = u ? o - S(Tr, a) : 0;
      if (0 >= e && 0 == c) return {
          line: a,
          ch: 0
      };
      for (var f, h = !1, d = 0, p = 0, m = s.length, v = Math.min(m, Math.ceil((e + .9 * c * xr.clientWidth) / i));;) {
        var y = n(v);
        if (!(e >= y && m > v)) {
          f = y, m = v;
          break
        }
        v = Math.min(m, Math.ceil(1.2 * v))
      }
      if (e > f) return {
          line: a,
          ch: m
      };
      for (v = Math.floor(.8 * m), y = n(v), e > y && (d = v, p = y);;) {
        if (1 >= m - d) {
          var x = f - e > e - p;
          return {
            line: a,
            ch: x ? d : m,
            after: x
          }
        }
        var C = Math.ceil((d + m) / 2),
          b = n(C);
        b > e ? (m = C, f = b, h && (f += 1e3)) : (d = C, p = b)
      }
    }

    function Rn(e) {
      var t = Pn(e, !0),
        n = H(pr);
      return {
        x: n.left + t.x,
        y: n.top + t.y,
        yBot: n.top + t.yBot
      }
    }

    function jn() {
      if (null == ci) {
        ci = K("pre");
        for (var e = 0; 49 > e; ++e) ci.appendChild(document.createTextNode("x")), ci.appendChild(K("br"));
        ci.appendChild(document.createTextNode("x"))
      }
      var t = ur.clientHeight;
      return t == ui ? si : (ui = t, Y(dr, ci.cloneNode(!0)), si = dr.firstChild.offsetHeight / 50 || 1, U(dr), si)
    }

    function Fn() {
      if (xr.clientWidth == di) return fi;
      di = xr.clientWidth;
      var e = K("span", "x"),
        t = K("pre", [e]);
      return Y(dr, t), fi = e.offsetWidth || 10
    }

    function Vn() {
      return pr.offsetTop
    }

    function Gn() {
      return pr.offsetLeft
    }

    function Kn(e, t) {
      var n, r, i = H(xr, !0);
      try {
        n = e.clientX, r = e.clientY
      } catch (e) {
        return null
      }
      if (!t && (n - i.left > xr.clientWidth || r - i.top > xr.clientHeight)) return null;
      var o = H(pr, !0);
      return Bn(n - o.left, r - o.top)
    }

    function Un(e) {
      function t() {
        if (ar.style.position = "relative", or.style.cssText = i, ut && (sr.scrollTop = r), Kt(), null != or.selectionStart) {
          clearTimeout(hi);
          var e = or.value = " " + (F(jr.from, jr.to) ? "" : or.value),
            t = 0;
          li = " ", or.selectionStart = 1, or.selectionEnd = e.length, hi = setTimeout(function n() {
              " " == li && 0 == or.selectionStart ? er(it.selectAll)(oi) : 10 > t++ ? hi = setTimeout(n, 500) : $t()
            }, 200)
        }
      }
      var n = Kn(e),
        r = sr.scrollTop;
      if (n && !dt) {
        (F(jr.from, jr.to) || V(n, jr.from) || !V(n, jr.to)) && er(cn)(n.line, n.ch);
        var i = or.style.cssText;
        if (ar.style.position = "absolute", or.style.cssText = "position: fixed; width: 30px; height: 30px; top: " + (e.clientY - 5) + "px; left: " + (e.clientX - 5) + "px; z-index: 1000; background: white; " + "border-width: 0; outline: none; overflow: hidden; opacity: .05; filter: alpha(opacity=5);", Zt(), $t(!0), F(jr.from, jr.to) && (or.value = li = " "), at) {
          W(e);
          var o = z(window, "mouseup", function () {
              o(), setTimeout(t, 20)
            }, !0)
        } else setTimeout(t, 50)
      }
    }

    function Yn() {
      clearInterval(br);
      var e = !0;
      fr.style.visibility = "", br = setInterval(function () {
          fr.style.visibility = (e = !e) ? "" : "hidden"
        }, nr.cursorBlinkRate)
    }

    function $n(e) {
      function t(e, t, n) {
        if (e.text) for (var r, i = e.styles, o = a ? 0 : e.text.length - 1, s = a ? 0 : i.length - 2, u = a ? i.length : -2; s != u; s += 2 * l) {
            var c = i[s];
            if (i[s + 1] == h) {
              for (var f = a ? 0 : c.length - 1, m = a ? c.length : -1; f != m; f += l, o += l) if (o >= t && n > o && p.test(r = c.charAt(f))) {
                  var g = pi[r];
                  if (">" == g.charAt(1) == a) d.push(r);
                  else {
                    if (d.pop() != g.charAt(0)) return {
                        pos: o,
                        match: !1
                    };
                    if (!d.length) return {
                        pos: o,
                        match: !0
                    }
                  }
                }
            } else o += l * c.length
        }
      }
      var n = jr.inverted ? jr.from : jr.to,
        r = g(n.line),
        i = n.ch - 1,
        o = i >= 0 && pi[r.text.charAt(i)] || pi[r.text.charAt(++i)];
      if (o) {
        for (var a = (o.charAt(0), ">" == o.charAt(1)), l = a ? 1 : -1, s = r.styles, u = i + 1, c = 0, f = s.length; f > c; c += 2) if (0 >= (u -= s[c].length)) {
            var h = s[c + 1];
            break
          }
        for (var d = [r.text.charAt(i)], p = /[(){}[\]]/, c = n.line, f = a ? Math.min(c + 100, Tr.size) : Math.max(-1, c - 100); c != f; c += l) {
          var r = g(c),
            m = c == n.line,
            v = t(r, m && a ? i + 1 : 0, m && !a ? i : r.text.length);
          if (v) break
        }
        v || (v = {
            pos: null,
            match: !1
          });
        var h = v.match ? "CodeMirror-matchingbracket" : "CodeMirror-nonmatchingbracket",
          y = An({
              line: n.line,
              ch: i
            }, {
              line: n.line,
              ch: i + 1
            }, h),
          x = null != v.pos && An({
              line: c,
              ch: v.pos
            }, {
              line: c,
              ch: v.pos + 1
            }, h),
          C = er(function () {
              y.clear(), x && x.clear()
            });
        e ? setTimeout(C, 800) : Rr = C
      }
    }

    function Zn(e) {
      for (var t, n, r = e, i = e - 40; r > i; --r) {
        if (0 == r) return 0;
        var o = g(r - 1);
        if (o.stateAfter) return r;
        var a = o.indentation(nr.tabSize);
        (null == n || t > a) && (n = r - 1, t = a)
      }
      return n
    }

    function Jn(e) {
      var t = Zn(e),
        n = t && g(t - 1).stateAfter;
      return n = n ? i(wr, n) : o(wr), Tr.iter(t, e, function (r) {
          r.process(wr, n, nr.tabSize), r.stateAfter = t == e - 1 || 0 == t % 5 ? i(wr, n) : null
        }), n
    }

    function qn() {
      if (!(Ar >= $r)) {
        var e = +new Date + nr.workTime,
          t = i(wr, Jn(Ar)),
          n = Ar;
        Tr.iter(Ar, $r, function (n) {
            return Ar >= Yr ? (n.highlight(wr, t, nr.tabSize), n.stateAfter = i(wr, t)) : (n.process(wr, t, nr.tabSize), n.stateAfter = 0 == Ar % 5 ? i(wr, t) : null), ++Ar, +new Date > e ? (Qn(nr.workDelay), !0) : void 0
          }), $r > n && Ar >= Yr && er(function () {
            Or.push({
                from: n,
                to: Ar
              })
          })()
      }
    }

    function Qn(e) {
      $r > Ar && Mr.set(e, qn)
    }

    function Xn() {
      Dr = zr = Ir = null, Or = [], Hr = !1, Br = []
    }

    function _n() {
      if (qr && Bt(), Qr && !nr.lineWrapping) {
        var e = hr.offsetWidth,
          t = Hn(Jr, Jr.text.length).left;
        st || (hr.style.left = t + "px", pr.style.minWidth = t + e + "px"), Qr = !1
      }
      var n, r;
      if (Hr) {
        var i = qt();
        n = Xt(i.x, i.y, i.x, i.yBot)
      }(Or.length || n && null != n.scrollTop) && (r = en(Or, !0, n && n.scrollTop)), r || (Hr && on(), Pr && rn()), n && Jt(), Hr && Yn(), Sr && (Dr === !0 || Dr !== !1 && Hr) && $t(zr), Hr && nr.matchBrackets && setTimeout(er(function () {
            Rr && (Rr(), Rr = null), F(jr.from, jr.to) && $n(!1)
          }), 20);
      var o = Hr,
        a = Br;
      Ir && nr.onChange && oi && nr.onChange(oi, Ir), o && nr.onCursorActivity && nr.onCursorActivity(oi);
      for (var l = 0; a.length > l; ++l) a[l](oi);
      r && nr.onUpdate && nr.onUpdate(oi)
    }

    function er(e) {
      return function () {
        mi++ || Xn();
        try {
          var t = e.apply(this, arguments)
        } finally {
          --mi || _n()
        }
        return t
      }
    }

    function tr(e) {
      ei.startCompound();
      try {
        return e()
      } finally {
        ei.endCompound()
      }
    }
    var nr = {}, rr = e.defaults;
    for (var ir in rr) rr.hasOwnProperty(ir) && (nr[ir] = (u && u.hasOwnProperty(ir) ? u : rr)[ir]);
    var or = K("textarea", null, null, "position: absolute; padding: 0; width: 1px; height: 1em");
    or.setAttribute("wrap", "off"), or.setAttribute("autocorrect", "off"), or.setAttribute("autocapitalize", "off");
    var ar = K("div", [or], null, "overflow: hidden; position: relative; width: 3px; height: 0px;"),
      lr = K("div", null, "CodeMirror-scrollbar-inner"),
      sr = K("div", [lr], "CodeMirror-scrollbar"),
      ur = K("div"),
      cr = K("div", null, null, "position: relative; z-index: -1"),
      fr = K("pre", " ", "CodeMirror-cursor"),
      hr = K("pre", " ", "CodeMirror-cursor", "visibility: hidden"),
      dr = K("div", null, null, "position: absolute; width: 100%; height: 0px; overflow: hidden; visibility: hidden;"),
      pr = K("div", [dr, fr, hr, cr, ur], null, "position: relative; z-index: 0"),
      mr = K("div", null, "CodeMirror-gutter-text"),
      gr = K("div", [mr], "CodeMirror-gutter"),
      vr = K("div", [gr, K("div", [pr], "CodeMirror-lines")], null, "position: relative"),
      yr = K("div", [vr], null, "position: relative"),
      xr = K("div", [yr], "CodeMirror-scroll");
    xr.setAttribute("tabIndex", "-1");
    var Cr = K("div", [ar, sr, xr], "CodeMirror" + (nr.lineWrapping ? " CodeMirror-wrap" : ""));
    a.appendChild ? a.appendChild(Cr) : a(Cr), kn(), Mn(), Q && (or.style.width = "0px"), ft || (xr.draggable = !0), pr.style.outline = "none", null != nr.tabindex && (or.tabIndex = nr.tabindex), nr.autofocus && Zt(), nr.gutter || nr.lineNumbers || (gr.style.display = "none"), mt && (ar.style.height = "1px", ar.style.position = "absolute"), gt ? (sr.style.zIndex = -2, sr.style.visibility = "hidden") : st && (sr.style.minWidth = "18px");
    var br, wr, Sr, kr = new O,
      Mr = new O,
      Tr = new x([new y([new v("")])]),
      Ar = 0;
    bn();
    var Er, Wr, Lr, Nr, Dr, zr, Or, Ir, Hr, Pr, Br, Rr, jr = {
        from: {
          line: 0,
          ch: 0
        },
        to: {
          line: 0,
          ch: 0
        },
        inverted: !1
      }, Fr = 0,
      Vr = !1,
      Gr = !1,
      Kr = !1,
      Ur = 0,
      Yr = 0,
      $r = 0,
      Zr = 0,
      Jr = g(0),
      qr = !1,
      Qr = !0,
      Xr = !1,
      _r = null;
    er(function () {
        I(nr.value || ""), Dr = !1
      })();
    var ei = new k;
    z(xr, "mousedown", er(rt)), z(xr, "dblclick", er(ht)), z(pr, "selectstart", A), at || z(xr, "contextmenu", Un), z(xr, "scroll", et), z(sr, "scroll", _), z(sr, "mousedown", function () {
        Sr && setTimeout(Zt, 0)
      });
    var ti = z(window, "resize", function () {
        Cr.parentNode ? en(!0) : ti()
      }, !0);
    z(or, "keyup", er(Et)), z(or, "input", Ut), z(or, "keydown", er(Tt)), z(or, "keypress", er(At)), z(or, "focus", Wt), z(or, "blur", Lt), nr.dragDrop && (z(xr, "dragstart", vt), z(xr, "dragenter", c), z(xr, "dragover", c), z(xr, "drop", er(pt))), z(xr, "paste", function () {
        Zt(), Ut()
      }), z(or, "paste", function () {
        Kr = !0, Ut()
      }), z(or, "cut", er(function () {
          nr.readOnly || jt("")
        })), mt && z(yr, "mouseup", function () {
        document.activeElement == or && or.blur(), Zt()
      });
    var ni;
    try {
      ni = document.activeElement == or
    } catch (ri) {}
    ni || nr.autofocus ? setTimeout(Wt, 20) : Lt();
    var ii, oi = Cr.CodeMirror = {
        getValue: $,
        setValue: er(I),
        getSelection: Gt,
        replaceSelection: er(jt),
        focus: function () {
          window.focus(), Zt(), Wt(), Ut()
        },
        setOption: function (e, t) {
          var n = nr[e];
          nr[e] = t, "mode" == e || "indentUnit" == e ? bn() : "readOnly" == e && "nocursor" == t ? (Lt(), or.blur()) : "readOnly" != e || t ? "theme" == e ? kn() : "lineWrapping" == e && n != t ? er(Sn)() : "tabSize" == e ? en(!0) : "keyMap" == e ? Mn() : "tabindex" == e && (or.tabIndex = t) : $t(!0), ("lineNumbers" == e || "gutter" == e || "firstLineNumber" == e || "theme" == e || "lineNumberFormatter" == e) && (wn(), en(!0))
        },
        getOption: function (e) {
          return nr[e]
        },
        getMode: function () {
          return wr
        },
        undo: er(zt),
        redo: er(Ot),
        indentLine: er(function (e, t) {
            "string" != typeof t && (t = null == t ? nr.smartIndent ? "smart" : "prev" : t ? "add" : "subtract"), m(e) && Cn(e, t)
          }),
        indentSelection: er(xn),
        historySize: function () {
          return {
            undo: ei.done.length,
            redo: ei.undone.length
          }
        },
        clearHistory: function () {
          ei = new k
        },
        setHistory: function (e) {
          ei = new k, ei.done = e.done, ei.undone = e.undone
        },
        getHistory: function () {
          function e(e) {
            for (var t, n = 0, r = []; e.length > n; ++n) {
              r.push(t = []);
              for (var i = 0, o = e[n]; o.length > i; ++i) {
                var a = [],
                  l = o[i];
                t.push({
                    start: l.start,
                    added: l.added,
                    old: a
                  });
                for (var s = 0; l.old.length > s; ++s) a.push(h(l.old[s]))
              }
            }
            return r
          }
          return {
            done: e(ei.done),
            undone: e(ei.undone)
          }
        },
        matchBrackets: er(function () {
            $n(!0)
          }),
        getTokenAt: er(function (e) {
            return e = hn(e), g(e.line).getTokenAt(wr, Jn(e.line), nr.tabSize, e.ch)
          }),
        getStateAfter: function (e) {
          return e = fn(null == e ? Tr.size - 1 : e), Jn(e + 1)
        },
        cursorCoords: function (e, t) {
          return null == e && (e = jr.inverted), this.charCoords(e ? jr.from : jr.to, t)
        },
        charCoords: function (e, t) {
          return e = hn(e), "local" == t ? Pn(e, !1) : "div" == t ? Pn(e, !0) : Rn(e)
        },
        coordsChar: function (e) {
          var t = H(pr);
          return Bn(e.x - t.left, e.y - t.top)
        },
        markText: er(An),
        setBookmark: En,
        findMarksAt: Wn,
        setMarker: er(Ln),
        clearMarker: er(Nn),
        setLineClass: er(zn),
        hideLine: er(function (e) {
            return On(e, !0)
          }),
        showLine: er(function (e) {
            return On(e, !1)
          }),
        onDeleteLine: function (e, t) {
          if ("number" == typeof e) {
            if (!m(e)) return null;
            e = g(e)
          }
          return (e.handlers || (e.handlers = [])).push(t), e
        },
        lineInfo: In,
        getViewport: function () {
          return {
            from: Yr,
            to: $r
          }
        },
        addWidget: function (e, t, n, r, i) {
          e = Pn(hn(e));
          var o = e.yBot,
            a = e.x;
          if (t.style.position = "absolute", yr.appendChild(t), "over" == r) o = e.y;
          else if ("near" == r) {
            var l = Math.max(xr.offsetHeight, Tr.height * jn()),
              s = Math.max(yr.clientWidth, pr.clientWidth) - Gn();
            e.yBot + t.offsetHeight > l && e.y > t.offsetHeight && (o = e.y - t.offsetHeight), a + t.offsetWidth > s && (a = s - t.offsetWidth)
          }
          t.style.top = o + Vn() + "px", t.style.left = t.style.right = "", "right" == i ? (a = yr.clientWidth - t.offsetWidth, t.style.right = "0px") : ("left" == i ? a = 0 : "middle" == i && (a = (yr.clientWidth - t.offsetWidth) / 2), t.style.left = a + Gn() + "px"), n && Qt(a, o, a + t.offsetWidth, o + t.offsetHeight)
        },
        lineCount: function () {
          return Tr.size
        },
        clipPos: hn,
        getCursor: function (e) {
          return null == e && (e = jr.inverted), G(e ? jr.from : jr.to)
        },
        somethingSelected: function () {
          return !F(jr.from, jr.to)
        },
        setCursor: er(function (e, t, n) {
            null == t && "number" == typeof e.line ? cn(e.line, e.ch, n) : cn(e, t, n)
          }),
        setSelection: er(function (e, t, n) {
            (n ? ln : sn)(hn(e), hn(t || e))
          }),
        getLine: function (e) {
          return m(e) ? g(e).text : void 0
        },
        getLineHandle: function (e) {
          return m(e) ? g(e) : void 0
        },
        setLine: er(function (e, t) {
            m(e) && Rt(t, {
                line: e,
                ch: 0
              }, {
                line: e,
                ch: g(e).text.length
              })
          }),
        removeLine: er(function (e) {
            m(e) && Rt("", {
                line: e,
                ch: 0
              }, hn({
                  line: e + 1,
                  ch: 0
                }))
          }),
        replaceRange: er(Rt),
        getRange: function (e, t, n) {
          return Vt(hn(e), hn(t), n)
        },
        triggerOnKeyDown: er(Tt),
        execCommand: function (e) {
          return it[e](oi)
        },
        moveH: er(pn),
        deleteH: er(mn),
        moveV: er(gn),
        toggleOverwrite: function () {
          Vr ? (Vr = !1, fr.className = fr.className.replace(" CodeMirror-overwrite", "")) : (Vr = !0, fr.className += " CodeMirror-overwrite")
        },
        posFromIndex: function (e) {
          var t, n = 0;
          return Tr.iter(0, Tr.size, function (r) {
              var i = r.text.length + 1;
              return i > e ? (t = e, !0) : (e -= i, ++n, void 0)
            }), hn({
              line: n,
              ch: t
            })
        },
        indexFromPos: function (e) {
          if (0 > e.line || 0 > e.ch) return 0;
          var t = e.ch;
          return Tr.iter(0, e.line, function (e) {
              t += e.text.length + 1
            }), t
        },
        scrollTo: function (e, t) {
          null != e && (xr.scrollLeft = e), null != t && (sr.scrollTop = xr.scrollTop = t), en([])
        },
        getScrollInfo: function () {
          return {
            x: xr.scrollLeft,
            y: sr.scrollTop,
            height: sr.scrollHeight,
            width: xr.scrollWidth
          }
        },
        setSize: function (e, t) {
          function n(e) {
            return e += "", /^\d+$/.test(e) ? e + "px" : e
          }
          null != e && (Cr.style.width = n(e)), null != t && (xr.style.height = n(t)), oi.refresh()
        },
        operation: function (e) {
          return er(e)()
        },
        compoundChange: function (e) {
          return tr(e)
        },
        refresh: function () {
          en(!0, null, Fr), sr.scrollHeight > Fr && (sr.scrollTop = Fr)
        },
        getInputField: function () {
          return or
        },
        getWrapperElement: function () {
          return Cr
        },
        getScrollerElement: function () {
          return xr
        },
        getGutterElement: function () {
          return gr
        }
      }, ai = null,
      li = "";
    Tn.prototype.clear = er(function () {
        for (var e, t, n = 0; this.lines.length > n; ++n) {
          var r = this.lines[n],
            i = l(r.markedSpans, this);
          null != i.from && (e = b(r)), null != i.to && (t = b(r)), r.markedSpans = s(r.markedSpans, i)
        }
        null != e && Or.push({
            from: e,
            to: t + 1
          }), this.lines.length = 0, this.explicitlyCleared = !0
      }), Tn.prototype.find = function () {
      for (var e, t, n = 0; this.lines.length > n; ++n) {
        var r = this.lines[n],
          i = l(r.markedSpans, this);
        if (null != i.from || null != i.to) {
          var o = b(r);
          null != i.from && (e = {
              line: o,
              ch: i.from
            }), null != i.to && (t = {
              line: o,
              ch: i.to
            })
        }
      }
      return "bookmark" == this.type ? e : e && {
        from: e,
        to: t
      }
    };
    var si, ui, ci, fi, hi, di = 0,
      pi = {
        "(": ")>",
        ")": "(<",
        "[": "]>",
        "]": "[<",
        "{": "}>",
        "}": "{<"
      }, mi = 0;
    for (var gi in tt) tt.propertyIsEnumerable(gi) && !oi.propertyIsEnumerable(gi) && (oi[gi] = tt[gi]);
    for (var vi = 0; nt.length > vi; ++vi) nt[vi](oi);
    return oi
  }

  function t(e) {
    return "string" == typeof e ? ot[e] : e
  }

  function n(e, n, r, i, o) {
    function a(n) {
      n = t(n);
      var r = n[e];
      if (r === !1) return o && o(), !0;
      if (null != r && i(r)) return !0;
      if (n.nofallthrough) return o && o(), !0;
      var l = n.fallthrough;
      if (null == l) return !1;
      if ("[object Array]" != Object.prototype.toString.call(l)) return a(l);
      for (var s = 0, u = l.length; u > s; ++s) if (a(l[s])) return !0;
      return !1
    }
    return n && a(n) ? !0 : a(r)
  }

  function r(e) {
    var t = kt[D(e, "keyCode")];
    return "Ctrl" == t || "Alt" == t || "Shift" == t || "Mod" == t
  }

  function i(e, t) {
    if (t === !0) return t;
    if (e.copyState) return e.copyState(t);
    var n = {};
    for (var r in t) {
      var i = t[r];
      i instanceof Array && (i = i.concat([])), n[r] = i
    }
    return n
  }

  function o(e, t, n) {
    return e.startState ? e.startState(t, n) : !0
  }

  function a(e, t) {
    this.pos = this.start = 0, this.string = e, this.tabSize = t || 8
  }

  function l(e, t) {
    if (e) for (var n = 0; e.length > n; ++n) {
        var r = e[n];
        if (r.marker == t) return r
    }
  }

  function s(e, t) {
    for (var n, r = 0; e.length > r; ++r) e[r] != t && (n || (n = [])).push(e[r]);
    return n
  }

  function u(e, t, n) {
    if (e) for (var r, i = 0; e.length > i; ++i) {
        var o = e[i],
          a = o.marker,
          l = null == o.from || (a.inclusiveLeft ? t >= o.from : t > o.from);
        if (l || "bookmark" == a.type && o.from == t && o.from != n) {
          var s = null == o.to || (a.inclusiveRight ? o.to >= t : o.to > t);
          (r || (r = [])).push({
              from: o.from,
              to: s ? null : o.to,
              marker: a
            })
        }
    }
    return r
  }

  function c(e, t) {
    if (e) for (var n, r = 0; e.length > r; ++r) {
        var i = e[r],
          o = i.marker,
          a = null == i.to || (o.inclusiveRight ? i.to >= t : i.to > t);
        if (a || "bookmark" == o.type && i.from == t) {
          var l = null == i.from || (o.inclusiveLeft ? t >= i.from : t > i.from);
          (n || (n = [])).push({
              from: l ? null : i.from - t,
              to: null == i.to ? null : i.to - t,
              marker: o
            })
        }
    }
    return n
  }

  function f(e, t, n, r, i) {
    if (!e && !t) return i;
    var o = u(e, n),
      a = c(t, r),
      s = 1 == i.length,
      f = R(i).length + (s ? n : 0);
    if (o) for (var h = 0; o.length > h; ++h) {
        var d = o[h];
        if (null == d.to) {
          var m = l(a, d.marker);
          m ? s && (d.to = null == m.to ? null : m.to + f) : d.to = n
        }
    }
    if (a) for (var h = 0; a.length > h; ++h) {
        var d = a[h];
        if (null != d.to && (d.to += f), null == d.from) {
          var m = l(o, d.marker);
          m || (d.from = f, s && (o || (o = [])).push(d))
        } else d.from += f, s && (o || (o = [])).push(d)
    }
    var g = [p(i[0], o)];
    if (!s) {
      var v, y = i.length - 2;
      if (y > 0 && o) for (var h = 0; o.length > h; ++h) null == o[h].to && (v || (v = [])).push({
              from: null,
              to: null,
              marker: o[h].marker
            });
      for (var h = 0; y > h; ++h) g.push(p(i[h + 1], v));
      g.push(p(R(i), a))
    }
    return g
  }

  function h(e) {
    return "string" == typeof e ? e : e.text
  }

  function d(e) {
    if ("string" == typeof e) return null;
    for (var t = e.markedSpans, n = null, r = 0; t.length > r; ++r) t[r].marker.explicitlyCleared ? n || (n = t.slice(0, r)) : n && n.push(t[r]);
    return n ? n.length ? n : null : t
  }

  function p(e, t) {
    return t ? {
      text: e,
      markedSpans: t
    } : e
  }

  function m(e) {
    var t = e.markedSpans;
    if (t) {
      for (var n = 0; t.length > n; ++n) {
        var r = t[n].marker.lines,
          i = J(r, e);
        r.splice(i, 1)
      }
      e.markedSpans = null
    }
  }

  function g(e, t) {
    if (t) {
      for (var n = 0; t.length > n; ++n) t[n].marker.lines.push(e);
      e.markedSpans = t
    }
  }

  function v(e, t) {
    this.text = e, this.height = 1, g(this, t)
  }

  function y(e) {
    this.lines = e, this.parent = null;
    for (var t = 0, n = e.length, r = 0; n > t; ++t) e[t].parent = this, r += e[t].height;
    this.height = r
  }

  function x(e) {
    this.children = e;
    for (var t = 0, n = 0, r = 0, i = e.length; i > r; ++r) {
      var o = e[r];
      t += o.chunkSize(), n += o.height, o.parent = this
    }
    this.size = t, this.height = n, this.parent = null
  }

  function C(e, t) {
    for (; !e.lines;) for (var n = 0;; ++n) {
        var r = e.children[n],
          i = r.chunkSize();
        if (i > t) {
          e = r;
          break
        }
        t -= i
    }
    return e.lines[t]
  }

  function b(e) {
    if (null == e.parent) return null;
    for (var t = e.parent, n = J(t.lines, e), r = t.parent; r; t = r, r = r.parent) {
      var i = 0;
      for (r.children.length; r.children[i] != t; ++i) n += r.children[i].chunkSize()
    }
    return n
  }

  function w(e, t) {
    var n = 0;
    e: do {
      for (var r = 0, i = e.children.length; i > r; ++r) {
        var o = e.children[r],
          a = o.height;
        if (a > t) {
          e = o;
          continue e
        }
        t -= a, n += o.chunkSize()
      }
      return n
    } while (!e.lines);
    for (var r = 0, i = e.lines.length; i > r; ++r) {
      var l = e.lines[r],
        s = l.height;
      if (s > t) break;
      t -= s
    }
    return n + r
  }

  function S(e, t) {
    var n = 0;
    e: do {
      for (var r = 0, i = e.children.length; i > r; ++r) {
        var o = e.children[r],
          a = o.chunkSize();
        if (a > t) {
          e = o;
          continue e
        }
        t -= a, n += o.height
      }
      return n
    } while (!e.lines);
    for (var r = 0; t > r; ++r) n += e.lines[r].height;
    return n
  }

  function k() {
    this.time = 0, this.done = [], this.undone = [], this.compound = 0, this.closed = !1
  }

  function M() {
    W(this)
  }

  function T(e) {
    return e.stop || (e.stop = M), e
  }

  function A(e) {
    e.preventDefault ? e.preventDefault() : e.returnValue = !1
  }

  function E(e) {
    e.stopPropagation ? e.stopPropagation() : e.cancelBubble = !0
  }

  function W(e) {
    A(e), E(e)
  }

  function L(e) {
    return e.target || e.srcElement
  }

  function N(e) {
    var t = e.which;
    return null == t && (1 & e.button ? t = 1 : 2 & e.button ? t = 3 : 4 & e.button && (t = 2)), X && e.ctrlKey && 1 == t && (t = 3), t
  }

  function D(e, t) {
    var n = e.override && e.override.hasOwnProperty(t);
    return n ? e.override[t] : e[t]
  }

  function z(e, t, n, r) {
    if ("function" == typeof e.addEventListener) {
      if (e.addEventListener(t, n, !1), r) return function () {
          e.removeEventListener(t, n, !1)
      }
    } else {
      var i = function (e) {
        n(e || window.event)
      };
      if (e.attachEvent("on" + t, i), r) return function () {
          e.detachEvent("on" + t, i)
      }
    }
  }

  function O() {
    this.id = null
  }

  function I(e, t, n) {
    null == t && (t = e.search(/[^\s\u00a0]/), -1 == t && (t = e.length));
    for (var r = 0, i = 0; t > r; ++r) "  " == e.charAt(r) ? i += n - i % n : ++i;
    return i
  }

  function H(e, t) {
    try {
      var n = e.getBoundingClientRect();
      n = {
        top: n.top,
        left: n.left
      }
    } catch (r) {
      n = {
        top: 0,
        left: 0
      }
    }
    if (!t) if (null == window.pageYOffset) {
        var i = document.documentElement || document.body.parentNode;
        null == i.scrollTop && (i = document.body), n.top += i.scrollTop, n.left += i.scrollLeft
      } else n.top += window.pageYOffset, n.left += window.pageXOffset;
    return n
  }

  function P(e) {
    return e.textContent || e.innerText || e.nodeValue || ""
  }

  function B(e) {
    for (; e >= bt.length;) bt.push(R(bt) + " ");
    return bt[e]
  }

  function R(e) {
    return e[e.length - 1]
  }

  function j(e) {
    Q ? (e.selectionStart = 0, e.selectionEnd = e.value.length) : e.select()
  }

  function F(e, t) {
    return e.line == t.line && e.ch == t.ch
  }

  function V(e, t) {
    return e.line < t.line || e.line == t.line && e.ch < t.ch
  }

  function G(e) {
    return {
      line: e.line,
      ch: e.ch
    }
  }

  function K(e, t, n, r) {
    var i = document.createElement(e);
    if (n && (i.className = n), r && (i.style.cssText = r), "string" == typeof t) $(i, t);
    else if (t) for (var o = 0; t.length > o; ++o) i.appendChild(t[o]);
    return i
  }

  function U(e) {
    return e.innerHTML = "", e
  }

  function Y(e, t) {
    U(e).appendChild(t)
  }

  function $(e, t) {
    ut ? (e.innerHTML = "", e.appendChild(document.createTextNode(t))) : e.textContent = t
  }

  function Z(e, t) {
    if (!t) return 0;
    if (!e) return t.length;
    for (var n = e.length, r = t.length; n >= 0 && r >= 0 && e.charAt(n) == t.charAt(r); --n, --r);
    return r + 1
  }

  function J(e, t) {
    if (e.indexOf) return e.indexOf(t);
    for (var n = 0, r = e.length; r > n; ++n) if (e[n] == t) return n;
    return -1
  }

  function q(e) {
    return /\w/.test(e) || e.toUpperCase() != e.toLowerCase() || /[\u4E00-\u9FA5]/.test(e)
  }
  e.defaults = {
    value: "",
    mode: null,
    theme: "default",
    indentUnit: 2,
    indentWithTabs: !1,
    smartIndent: !0,
    tabSize: 4,
    keyMap: "default",
    extraKeys: null,
    electricChars: !0,
    autoClearEmptyLines: !1,
    onKeyEvent: null,
    onDragEvent: null,
    lineWrapping: !1,
    lineNumbers: !1,
    gutter: !1,
    fixedGutter: !1,
    firstLineNumber: 1,
    readOnly: !1,
    dragDrop: !0,
    onChange: null,
    onCursorActivity: null,
    onViewportChange: null,
    onGutterClick: null,
    onUpdate: null,
    onFocus: null,
    onBlur: null,
    onScroll: null,
    matchBrackets: !1,
    cursorBlinkRate: 530,
    workTime: 100,
    workDelay: 200,
    pollInterval: 100,
    undoDepth: 40,
    tabindex: null,
    autofocus: null,
    lineNumberFormatter: function (e) {
      return e
    }
  };
  var Q = /AppleWebKit/.test(navigator.userAgent) && /Mobile\/\w+/.test(navigator.userAgent),
    X = Q || /Mac/.test(navigator.platform);
  /Win/.test(navigator.platform);
  var _ = e.modes = {}, et = e.mimeModes = {};
  e.defineMode = function (t, n) {
    if (e.defaults.mode || "null" == t || (e.defaults.mode = t), arguments.length > 2) {
      n.dependencies = [];
      for (var r = 2; arguments.length > r; ++r) n.dependencies.push(arguments[r])
    }
    _[t] = n
  }, e.defineMIME = function (e, t) {
    et[e] = t
  }, e.resolveMode = function (t) {
    if ("string" == typeof t && et.hasOwnProperty(t)) t = et[t];
    else if ("string" == typeof t && /^[\w\-]+\/[\w\-]+\+xml$/.test(t)) return e.resolveMode("application/xml");
    return "string" == typeof t ? {
      name: t
    } : t || {
      name: "null"
    }
  }, e.getMode = function (t, n) {
    var n = e.resolveMode(n),
      r = _[n.name];
    if (!r) return e.getMode(t, "text/plain");
    var i = r(t, n);
    if (rt.hasOwnProperty(n.name)) {
      var o = rt[n.name];
      for (var a in o) o.hasOwnProperty(a) && (i[a] = o[a])
    }
    return i.name = n.name, i
  }, e.listModes = function () {
    var e = [];
    for (var t in _) _.propertyIsEnumerable(t) && e.push(t);
    return e
  }, e.listMIMEs = function () {
    var e = [];
    for (var t in et) et.propertyIsEnumerable(t) && e.push({
          mime: t,
          mode: et[t]
        });
    return e
  };
  var tt = e.extensions = {};
  e.defineExtension = function (e, t) {
    tt[e] = t
  };
  var nt = [];
  e.defineInitHook = function (e) {
    nt.push(e)
  };
  var rt = e.modeExtensions = {};
  e.extendMode = function (e, t) {
    var n = rt.hasOwnProperty(e) ? rt[e] : rt[e] = {};
    for (var r in t) t.hasOwnProperty(r) && (n[r] = t[r])
  };
  var it = e.commands = {
    selectAll: function (e) {
      e.setSelection({
          line: 0,
          ch: 0
        }, {
          line: e.lineCount() - 1
        })
    },
    killLine: function (e) {
      var t = e.getCursor(!0),
        n = e.getCursor(!1),
        r = !F(t, n);
      r || e.getLine(t.line).length != t.ch ? e.replaceRange("", t, r ? n : {
          line: t.line
        }) : e.replaceRange("", t, {
          line: t.line + 1,
          ch: 0
        })
    },
    deleteLine: function (e) {
      var t = e.getCursor().line;
      e.replaceRange("", {
          line: t,
          ch: 0
        }, {
          line: t
        })
    },
    undo: function (e) {
      e.undo()
    },
    redo: function (e) {
      e.redo()
    },
    goDocStart: function (e) {
      e.setCursor(0, 0, !0)
    },
    goDocEnd: function (e) {
      e.setSelection({
          line: e.lineCount() - 1
        }, null, !0)
    },
    goLineStart: function (e) {
      e.setCursor(e.getCursor().line, 0, !0)
    },
    goLineStartSmart: function (e) {
      var t = e.getCursor(),
        n = e.getLine(t.line),
        r = Math.max(0, n.search(/\S/));
      e.setCursor(t.line, r >= t.ch && t.ch ? 0 : r, !0)
    },
    goLineEnd: function (e) {
      e.setSelection({
          line: e.getCursor().line
        }, null, !0)
    },
    goLineUp: function (e) {
      e.moveV(-1, "line")
    },
    goLineDown: function (e) {
      e.moveV(1, "line")
    },
    goPageUp: function (e) {
      e.moveV(-1, "page")
    },
    goPageDown: function (e) {
      e.moveV(1, "page")
    },
    goCharLeft: function (e) {
      e.moveH(-1, "char")
    },
    goCharRight: function (e) {
      e.moveH(1, "char")
    },
    goColumnLeft: function (e) {
      e.moveH(-1, "column")
    },
    goColumnRight: function (e) {
      e.moveH(1, "column")
    },
    goWordLeft: function (e) {
      e.moveH(-1, "word")
    },
    goWordRight: function (e) {
      e.moveH(1, "word")
    },
    delCharLeft: function (e) {
      e.deleteH(-1, "char")
    },
    delCharRight: function (e) {
      e.deleteH(1, "char")
    },
    delWordLeft: function (e) {
      e.deleteH(-1, "word")
    },
    delWordRight: function (e) {
      e.deleteH(1, "word")
    },
    indentAuto: function (e) {
      e.indentSelection("smart")
    },
    indentMore: function (e) {
      e.indentSelection("add")
    },
    indentLess: function (e) {
      e.indentSelection("subtract")
    },
    insertTab: function (e) {
      e.replaceSelection("  ", "end")
    },
    defaultTab: function (e) {
      e.somethingSelected() ? e.indentSelection("add") : e.replaceSelection(" ", "end")
    },
    transposeChars: function (e) {
      var t = e.getCursor(),
        n = e.getLine(t.line);
      t.ch > 0 && t.ch < n.length - 1 && e.replaceRange(n.charAt(t.ch) + n.charAt(t.ch - 1), {
          line: t.line,
          ch: t.ch - 1
        }, {
          line: t.line,
          ch: t.ch + 1
        })
    },
    newlineAndIndent: function (e) {
      e.replaceSelection("\n", "end"), e.indentLine(e.getCursor().line)
    },
    toggleOverwrite: function (e) {
      e.toggleOverwrite()
    }
  }, ot = e.keyMap = {};
  ot.basic = {
    Left: "goCharLeft",
    Right: "goCharRight",
    Up: "goLineUp",
    Down: "goLineDown",
    End: "goLineEnd",
    Home: "goLineStartSmart",
    PageUp: "goPageUp",
    PageDown: "goPageDown",
    Delete: "delCharRight",
    Backspace: "delCharLeft",
    Tab: "defaultTab",
    "Shift-Tab": "indentAuto",
    Enter: "newlineAndIndent",
    Insert: "toggleOverwrite"
  }, ot.pcDefault = {
    "Ctrl-A": "selectAll",
    "Ctrl-D": "deleteLine",
    "Ctrl-Z": "undo",
    "Shift-Ctrl-Z": "redo",
    "Ctrl-Y": "redo",
    "Ctrl-Home": "goDocStart",
    "Alt-Up": "goDocStart",
    "Ctrl-End": "goDocEnd",
    "Ctrl-Down": "goDocEnd",
    "Ctrl-Left": "goWordLeft",
    "Ctrl-Right": "goWordRight",
    "Alt-Left": "goLineStart",
    "Alt-Right": "goLineEnd",
    "Ctrl-Backspace": "delWordLeft",
    "Ctrl-Delete": "delWordRight",
    "Ctrl-S": "save",
    "Ctrl-F": "find",
    "Ctrl-G": "findNext",
    "Shift-Ctrl-G": "findPrev",
    "Shift-Ctrl-F": "replace",
    "Shift-Ctrl-R": "replaceAll",
    "Ctrl-[": "indentLess",
    "Ctrl-]": "indentMore",
    fallthrough: "basic"
  }, ot.macDefault = {
    "Cmd-A": "selectAll",
    "Cmd-D": "deleteLine",
    "Cmd-Z": "undo",
    "Shift-Cmd-Z": "redo",
    "Cmd-Y": "redo",
    "Cmd-Up": "goDocStart",
    "Cmd-End": "goDocEnd",
    "Cmd-Down": "goDocEnd",
    "Alt-Left": "goWordLeft",
    "Alt-Right": "goWordRight",
    "Cmd-Left": "goLineStart",
    "Cmd-Right": "goLineEnd",
    "Alt-Backspace": "delWordLeft",
    "Ctrl-Alt-Backspace": "delWordRight",
    "Alt-Delete": "delWordRight",
    "Cmd-S": "save",
    "Cmd-F": "find",
    "Cmd-G": "findNext",
    "Shift-Cmd-G": "findPrev",
    "Cmd-Alt-F": "replace",
    "Shift-Cmd-Alt-F": "replaceAll",
    "Cmd-[": "indentLess",
    "Cmd-]": "indentMore",
    fallthrough: ["basic", "emacsy"]
  }, ot["default"] = X ? ot.macDefault : ot.pcDefault, ot.emacsy = {
    "Ctrl-F": "goCharRight",
    "Ctrl-B": "goCharLeft",
    "Ctrl-P": "goLineUp",
    "Ctrl-N": "goLineDown",
    "Alt-F": "goWordRight",
    "Alt-B": "goWordLeft",
    "Ctrl-A": "goLineStart",
    "Ctrl-E": "goLineEnd",
    "Ctrl-V": "goPageUp",
    "Shift-Ctrl-V": "goPageDown",
    "Ctrl-D": "delCharRight",
    "Ctrl-H": "delCharLeft",
    "Alt-D": "delWordRight",
    "Alt-Backspace": "delWordLeft",
    "Ctrl-K": "killLine",
    "Ctrl-T": "transposeChars"
  }, e.isModifierKey = r, e.fromTextArea = function (t, n) {
    function r() {
      t.value = u.getValue()
    }
    if (n || (n = {}), n.value = t.value, !n.tabindex && t.tabindex && (n.tabindex = t.tabindex), null == n.autofocus) {
      var i = document.body;
      try {
        i = document.activeElement
      } catch (o) {}
      n.autofocus = i == t || null != t.getAttribute("autofocus") && i == document.body
    }
    if (t.form) {
      var a = z(t.form, "submit", r, !0);
      if ("function" == typeof t.form.submit) {
        var l = t.form.submit;
        t.form.submit = function s() {
          r(), t.form.submit = l, t.form.submit(), t.form.submit = s
        }
      }
    }
    t.style.display = "none";
    var u = e(function (e) {
        t.parentNode.insertBefore(e, t.nextSibling)
      }, n);
    return u.save = r, u.getTextArea = function () {
      return t
    }, u.toTextArea = function () {
      r(), t.parentNode.removeChild(u.getWrapperElement()), t.style.display = "", t.form && (a(), "function" == typeof t.form.submit && (t.form.submit = l))
    }, u
  };
  var at = /gecko\/\d{7}/i.test(navigator.userAgent),
    lt = /MSIE \d/.test(navigator.userAgent),
    st = /MSIE [1-7]\b/.test(navigator.userAgent),
    ut = /MSIE [1-8]\b/.test(navigator.userAgent),
    ct = lt && 5 == document.documentMode,
    ft = /WebKit\//.test(navigator.userAgent),
    ht = /Chrome\//.test(navigator.userAgent),
    dt = /Opera\//.test(navigator.userAgent),
    pt = /Apple Computer/.test(navigator.vendor),
    mt = /KHTML\//.test(navigator.userAgent),
    gt = /Mac OS X 10\D([7-9]|\d\d)\D/.test(navigator.userAgent);
  e.copyState = i, e.startState = o, e.innerMode = function (e, t) {
    for (; e.innerMode;) {
      var n = e.innerMode(t);
      t = n.state, e = n.mode
    }
    return n || {
      mode: e,
      state: t
    }
  }, a.prototype = {
    eol: function () {
      return this.pos >= this.string.length
    },
    sol: function () {
      return 0 == this.pos
    },
    peek: function () {
      return this.string.charAt(this.pos) || void 0
    },
    next: function () {
      return this.pos < this.string.length ? this.string.charAt(this.pos++) : void 0
    },
    eat: function (e) {
      var t = this.string.charAt(this.pos);
      if ("string" == typeof e) var n = t == e;
      else var n = t && (e.test ? e.test(t) : e(t));
      return n ? (++this.pos, t) : void 0
    },
    eatWhile: function (e) {
      for (var t = this.pos; this.eat(e););
      return this.pos > t
    },
    eatSpace: function () {
      for (var e = this.pos;
        /[\s\u00a0]/.test(this.string.charAt(this.pos));)++this.pos;
      return this.pos > e
    },
    skipToEnd: function () {
      this.pos = this.string.length
    },
    skipTo: function (e) {
      var t = this.string.indexOf(e, this.pos);
      return t > -1 ? (this.pos = t, !0) : void 0
    },
    backUp: function (e) {
      this.pos -= e
    },
    column: function () {
      return I(this.string, this.start, this.tabSize)
    },
    indentation: function () {
      return I(this.string, null, this.tabSize)
    },
    match: function (e, t, n) {
      if ("string" != typeof e) {
        var r = this.string.slice(this.pos).match(e);
        return r && r.index > 0 ? null : (r && t !== !1 && (this.pos += r[0].length), r)
      }
      var i = function (e) {
        return n ? e.toLowerCase() : e
      };
      return i(this.string).indexOf(i(e), this.pos) == this.pos ? (t !== !1 && (this.pos += e.length), !0) : void 0
    },
    current: function () {
      return this.string.slice(this.start, this.pos)
    }
  }, e.StringStream = a;
  var vt = " ";
  at || lt && !st ? vt = "" : dt && (vt = ""), v.prototype = {
    update: function (e, t) {
      this.text = e, this.stateAfter = this.styles = null, m(this), g(this, t)
    },
    highlight: function (e, t, n) {
      var r = new a(this.text, n),
        i = this.styles || (this.styles = []),
        o = i.length = 0;
      for ("" == this.text && e.blankLine && e.blankLine(t); !r.eol();) {
        var l = e.token(r, t),
          s = r.current();
        if (r.start = r.pos, o && i[o - 1] == l ? i[o - 2] += s : s && (i[o++] = s, i[o++] = l), r.pos > 5e3) {
          i[o++] = this.text.slice(r.pos), i[o++] = null;
          break
        }
      }
    },
    process: function (e, t, n) {
      var r = new a(this.text, n);
      for ("" == this.text && e.blankLine && e.blankLine(t); !r.eol() && 5e3 >= r.pos;) e.token(r, t), r.start = r.pos
    },
    getTokenAt: function (e, t, n, r) {
      for (var i = this.text, o = new a(i, n); r > o.pos && !o.eol();) {
        o.start = o.pos;
        var l = e.token(o, t)
      }
      return {
        start: o.start,
        end: o.pos,
        string: o.current(),
        className: l || null,
        state: t
      }
    },
    indentation: function (e) {
      return I(this.text, null, e)
    },
    getContent: function (e, t, n) {
      function r(t, n, r) {
        if (n) {
          if (o && lt && " " == n.charAt(0) && (n = " " + n.slice(1)), o = !1, l.test(n)) for (var i = document.createDocumentFragment(), s = 0;;) {
              l.lastIndex = s;
              var u = l.exec(n),
                c = u ? u.index - s : n.length - s;
              if (c && (i.appendChild(document.createTextNode(n.slice(s, s + c))), a += c), !u) break;
              if (s += c + 1, " " == u[0]) {
                var f = e - a % e;
                i.appendChild(K("span", B(f), "cm-tab")), a += f
              } else {
                var h = K("span", "", "cm-invalidchar");
                h.title = "\\u" + u[0].charCodeAt(0).toString(16), i.appendChild(h), a += 1
              }
          } else {
            a += n.length;
            var i = document.createTextNode(n)
          }
          r ? t.appendChild(K("span", [i], r)) : t.appendChild(i)
        }
      }

      function i(e) {
        return e ? "cm-" + e.replace(/ +/g, " cm-") : null
      }
      var o = !0,
        a = 0,
        l = /[\t\u0000-\u0019\u200b\u2028\u2029\uFEFF]/g,
        s = K("pre"),
        u = r;
      if (null != t) {
        var c = 0,
          f = s.anchor = K("span");
        u = function (e, i, o) {
          var a = i.length;
          if (t >= c && c + a > t) {
            var l = t - c;
            if (l && (r(e, i.slice(0, l), o), n)) {
              var s = i.slice(l - 1, l + 1);
              Ct.test(s) ? e.appendChild(K("wbr")) : !st && /\w\w/.test(s) && e.appendChild(document.createTextNode(""))
            }
            e.appendChild(f), r(f, dt ? i.slice(l, l + 1) : i.slice(l), o), dt && r(e, i.slice(l + 1), o), t--, c += a
          } else c += a, r(e, i, o), c == t && c == m ? ($(f, vt), e.appendChild(f)) : c > t + 10 && /\s/.test(i) && (u = function () {})
        }
      }
      var h = this.styles,
        d = this.text,
        p = this.markedSpans,
        m = d.length;
      if (d || null != t) if (p && p.length) {
          p.sort(function (e, t) {
              return e.from - t.from
            });
          for (var g, v = 0, y = 0, x = "", C = p[0].from || 0, b = [], w = 0, S = function () {
                for (var e; p.length > w && ((e = p[w]).from == v || null == e.from);) "range" == e.marker.type && b.push(e), ++w;
                C = p.length > w ? p[w].from : 1 / 0;
                for (var t = 0; b.length > t; ++t) {
                  var n = b[t].to;
                  null == n && (n = 1 / 0), n == v ? b.splice(t--, 1) : C = Math.min(n, C)
                }
              }; m > v;) {
            C == v && S();
            for (var k = Math.min(m, C);;) {
              if (x) {
                for (var M = v + x.length, T = g, A = 0; b.length > A; ++A) {
                  var E = b[A];
                  T = (T ? T + " " : "") + E.marker.style, E.marker.endStyle && E.to === Math.min(M, k) && (T += " " + E.marker.endStyle), E.marker.startStyle && E.from === v && (T += " " + E.marker.startStyle)
                }
                if (u(s, M > k ? x.slice(0, k - v) : x, T), M >= k) {
                  x = x.slice(k - v), v = k;
                  break
                }
                v = M
              }
              x = h[y++], g = i(h[y++])
            }
          }
        } else for (var y = 0, W = 0; m > W; y += 2) {
            var L = h[y],
              g = h[y + 1],
              N = L.length;
            W + N > m && (L = L.slice(0, m - W)), W += N, u(s, L, i(g))
        } else u(s, " ");
      return s
    },
    cleanUp: function () {
      this.parent = null, m(this)
    }
  }, y.prototype = {
    chunkSize: function () {
      return this.lines.length
    },
    remove: function (e, t, n) {
      for (var r = e, i = e + t; i > r; ++r) {
        var o = this.lines[r];
        if (this.height -= o.height, o.cleanUp(), o.handlers) for (var a = 0; o.handlers.length > a; ++a) n.push(o.handlers[a])
      }
      this.lines.splice(e, t)
    },
    collapse: function (e) {
      e.splice.apply(e, [e.length, 0].concat(this.lines))
    },
    insertHeight: function (e, t, n) {
      this.height += n, this.lines = this.lines.slice(0, e).concat(t).concat(this.lines.slice(e));
      for (var r = 0, i = t.length; i > r; ++r) t[r].parent = this
    },
    iterN: function (e, t, n) {
      for (var r = e + t; r > e; ++e) if (n(this.lines[e])) return !0
    }
  }, x.prototype = {
    chunkSize: function () {
      return this.size
    },
    remove: function (e, t, n) {
      this.size -= t;
      for (var r = 0; this.children.length > r; ++r) {
        var i = this.children[r],
          o = i.chunkSize();
        if (o > e) {
          var a = Math.min(t, o - e),
            l = i.height;
          if (i.remove(e, a, n), this.height -= l - i.height, o == a && (this.children.splice(r--, 1), i.parent = null), 0 == (t -= a)) break;
          e = 0
        } else e -= o
      }
      if (25 > this.size - t) {
        var s = [];
        this.collapse(s), this.children = [new y(s)], this.children[0].parent = this
      }
    },
    collapse: function (e) {
      for (var t = 0, n = this.children.length; n > t; ++t) this.children[t].collapse(e)
    },
    insert: function (e, t) {
      for (var n = 0, r = 0, i = t.length; i > r; ++r) n += t[r].height;
      this.insertHeight(e, t, n)
    },
    insertHeight: function (e, t, n) {
      this.size += t.length, this.height += n;
      for (var r = 0, i = this.children.length; i > r; ++r) {
        var o = this.children[r],
          a = o.chunkSize();
        if (a >= e) {
          if (o.insertHeight(e, t, n), o.lines && o.lines.length > 50) {
            for (; o.lines.length > 50;) {
              var l = o.lines.splice(o.lines.length - 25, 25),
                s = new y(l);
              o.height -= s.height, this.children.splice(r + 1, 0, s), s.parent = this
            }
            this.maybeSpill()
          }
          break
        }
        e -= a
      }
    },
    maybeSpill: function () {
      if (!(10 >= this.children.length)) {
        var e = this;
        do {
          var t = e.children.splice(e.children.length - 5, 5),
            n = new x(t);
          if (e.parent) {
            e.size -= n.size, e.height -= n.height;
            var r = J(e.parent.children, e);
            e.parent.children.splice(r + 1, 0, n)
          } else {
            var i = new x(e.children);
            i.parent = e, e.children = [i, n], e = i
          }
          n.parent = e.parent
        } while (e.children.length > 10);
        e.parent.maybeSpill()
      }
    },
    iter: function (e, t, n) {
      this.iterN(e, t - e, n)
    },
    iterN: function (e, t, n) {
      for (var r = 0, i = this.children.length; i > r; ++r) {
        var o = this.children[r],
          a = o.chunkSize();
        if (a > e) {
          var l = Math.min(t, a - e);
          if (o.iterN(e, l, n)) return !0;
          if (0 == (t -= l)) break;
          e = 0
        } else e -= a
      }
    }
  }, k.prototype = {
    addChange: function (e, t, n) {
      this.undone.length = 0;
      var r = +new Date,
        i = R(this.done),
        o = i && R(i),
        a = r - this.time;
      if (i && !this.closed && this.compound) i.push({
            start: e,
            added: t,
            old: n
          });
      else if (a > 400 || !o || this.closed || o.start > e + n.length || e > o.start + o.added) this.done.push([{
              start: e,
              added: t,
              old: n
            }
          ]), this.closed = !1;
      else {
        for (var l = Math.max(0, o.start - e), s = Math.max(0, e + n.length - (o.start + o.added)), u = l; u > 0; --u) o.old.unshift(n[u - 1]);
        for (var u = s; u > 0; --u) o.old.push(n[n.length - u]);
        l && (o.start = e), o.added += t - (n.length - l - s)
      }
      this.time = r
    },
    startCompound: function () {
      this.compound++ || (this.closed = !0)
    },
    endCompound: function () {
      --this.compound || (this.closed = !0)
    }
  }, e.e_stop = W, e.e_preventDefault = A, e.e_stopPropagation = E, e.connect = z, O.prototype = {
    set: function (e, t) {
      clearTimeout(this.id), this.id = setTimeout(t, e)
    }
  };
  var yt = e.Pass = {
    toString: function () {
      return "CodeMirror.Pass"
    }
  }, xt = function () {
      if (ut) return !1;
      var e = K("div");
      return "draggable" in e || "dragDrop" in e
    }();
  (function () {
      var e = K("textarea");
      return e.value = "foo\nbar", e.value.indexOf("\r") > -1 ? "\r\n" : "\n"
    })();
  var Ct = /^$/;
  at ? Ct = /$'/ : pt ? Ct = /\-[^ \-?]|\?[^ !'\"\),.\-\/:;\?\]\}]/ : ht && (Ct = /\-[^ \-\.?]|\?[^ \-\.?\]\}:;!'\"\),\/]|[\.!\"#&%\)*+,:;=>\]|\}~][\(\{\[<]|\$'/);
  var bt = [""],
    wt = 3 != "\n\nb".split(/\n/).length ? function (e) {
      for (var t = 0, n = [], r = e.length; r >= t;) {
        var i = e.indexOf("\n", t); - 1 == i && (i = e.length);
        var o = e.slice(t, "\r" == e.charAt(i - 1) ? i - 1 : i),
          a = o.indexOf("\r"); - 1 != a ? (n.push(o.slice(0, a)), t += a + 1) : (n.push(o), t = i + 1)
      }
      return n
    } : function (e) {
      return e.split(/\r\n?|\n/)
    };
  e.splitLines = wt;
  var St = window.getSelection ? function (e) {
      try {
        return e.selectionStart != e.selectionEnd
      } catch (t) {
        return !1
      }
    } : function (e) {
      try {
        var t = e.ownerDocument.selection.createRange()
      } catch (n) {}
      return t && t.parentElement() == e ? 0 != t.compareEndPoints("StartToEnd", t) : !1
    };
  e.defineMode("null", function () {
      return {
        token: function (e) {
          e.skipToEnd()
        }
      }
    }), e.defineMIME("text/plain", "null");
  var kt = {
    3: "Enter",
    8: "Backspace",
    9: "Tab",
    13: "Enter",
    16: "Shift",
    17: "Ctrl",
    18: "Alt",
    19: "Pause",
    20: "CapsLock",
    27: "Esc",
    32: "Space",
    33: "PageUp",
    34: "PageDown",
    35: "End",
    36: "Home",
    37: "Left",
    38: "Up",
    39: "Right",
    40: "Down",
    44: "PrintScrn",
    45: "Insert",
    46: "Delete",
    59: ";",
    91: "Mod",
    92: "Mod",
    93: "Mod",
    109: "-",
    107: "=",
    127: "Delete",
    186: ";",
    187: "=",
    188: ",",
    189: "-",
    190: ".",
    191: "/",
    192: "`",
    219: "[",
    220: "\\",
    221: "]",
    222: "'",
    63276: "PageUp",
    63277: "PageDown",
    63275: "End",
    63273: "Home",
    63234: "Left",
    63232: "Up",
    63235: "Right",
    63233: "Down",
    63302: "Insert",
    63272: "Delete"
  };
  return e.keyNames = kt,

  function () {
    for (var e = 0; 10 > e; e++) kt[e + 48] = e + "";
    for (var e = 65; 90 >= e; e++) kt[e] = String.fromCharCode(e);
    for (var e = 1; 12 >= e; e++) kt[e + 111] = kt[e + 63235] = "F" + e
  }(), e.version = "2.35", e
}(), CodeMirror.defineMode("javascript", function (e, t) {
    function n(e, t, n) {
      return t.tokenize = n, n(e, t)
    }

    function r(e, t) {
      for (var n, r = !1; null != (n = e.next());) {
        if (n == t && !r) return !1;
        r = !r && "\\" == n
      }
      return r
    }

    function i(e, t, n) {
      return B = e, R = n, t
    }

    function o(e, t) {
      var o = e.next();
      if ('"' == o || "'" == o) return n(e, t, a(o));
      if (/[\[\]{}\(\),;\:\.]/.test(o)) return i(o);
      if ("0" == o && e.eat(/x/i)) return e.eatWhile(/[\da-f]/i), i("number", "number");
      if (/\d/.test(o) || "-" == o && e.eat(/\d/)) return e.match(/^\d*(?:\.\d*)?(?:[eE][+\-]?\d+)?/), i("number", "number");
      if ("/" == o) return e.eat("*") ? n(e, t, l) : e.eat("/") ? (e.skipToEnd(), i("comment", "comment")) : "operator" == t.lastType || "keyword c" == t.lastType || /^[\[{}\(,;:]$/.test(t.lastType) ? (r(e, "/"), e.eatWhile(/[gimy]/), i("regexp", "string-2")) : (e.eatWhile(K), i("operator", null, e.current()));
      if ("#" == o) return e.skipToEnd(), i("error", "error");
      if (K.test(o)) return e.eatWhile(K), i("operator", null, e.current());
      e.eatWhile(/[\w\$_]/);
      var s = e.current(),
        u = G.propertyIsEnumerable(s) && G[s];
      return u && "." != t.lastType ? i(u.type, u.style, s) : i("variable", "variable", s)
    }

    function a(e) {
      return function (t, n) {
        return r(t, e) || (n.tokenize = o), i("string", "string")
      }
    }

    function l(e, t) {
      for (var n, r = !1; n = e.next();) {
        if ("/" == n && r) {
          t.tokenize = o;
          break
        }
        r = "*" == n
      }
      return i("comment", "comment")
    }

    function s(e, t, n, r, i, o) {
      this.indented = e, this.column = t, this.type = n, this.prev = i, this.info = o, null != r && (this.align = r)
    }

    function u(e, t) {
      for (var n = e.localVars; n; n = n.next) if (n.name == t) return !0
    }

    function c(e, t, n, r, i) {
      var o = e.cc;
      for (Y.state = e, Y.stream = i, Y.marked = null, Y.cc = o, e.lexical.hasOwnProperty("align") || (e.lexical.align = !0);;) {
        var a = o.length ? o.pop() : F ? C : x;
        if (a(n, r)) {
          for (; o.length && o[o.length - 1].lex;) o.pop()();
          return Y.marked ? Y.marked : "variable" == n && u(e, r) ? "variable-2" : t
        }
      }
    }

    function f() {
      for (var e = arguments.length - 1; e >= 0; e--) Y.cc.push(arguments[e])
    }

    function h() {
      return f.apply(null, arguments), !0
    }

    function d(e) {
      var t = Y.state;
      if (t.context) {
        Y.marked = "def";
        for (var n = t.localVars; n; n = n.next) if (n.name == e) return;
        t.localVars = {
          name: e,
          next: t.localVars
        }
      }
    }

    function p() {
      Y.state.context = {
        prev: Y.state.context,
        vars: Y.state.localVars
      }, Y.state.localVars = $
    }

    function m() {
      Y.state.localVars = Y.state.context.vars, Y.state.context = Y.state.context.prev
    }

    function g(e, t) {
      var n = function () {
        var n = Y.state;
        n.lexical = new s(n.indented, Y.stream.column(), e, null, n.lexical, t)
      };
      return n.lex = !0, n
    }

    function v() {
      var e = Y.state;
      e.lexical.prev && (")" == e.lexical.type && (e.indented = e.lexical.indented), e.lexical = e.lexical.prev)
    }

    function y(e) {
      return function (t) {
        return t == e ? h() : ";" == e ? f() : h(arguments.callee)
      }
    }

    function x(e) {
      return "var" == e ? h(g("vardef"), L, y(";"), v) : "keyword a" == e ? h(g("form"), C, x, v) : "keyword b" == e ? h(g("form"), x, v) : "{" == e ? h(g("}"), A, v) : ";" == e ? h() : "function" == e ? h(H) : "for" == e ? h(g("form"), y("("), g(")"), D, y(")"), v, x, v) : "variable" == e ? h(g("stat"), S) : "switch" == e ? h(g("form"), C, g("}", "switch"), y("{"), A, v, v) : "case" == e ? h(C, y(":")) : "default" == e ? h(y(":")) : "catch" == e ? h(g("form"), p, y("("), P, y(")"), x, v, m) : f(g("stat"), C, y(";"), v)
    }

    function C(e) {
      return U.hasOwnProperty(e) ? h(w) : "function" == e ? h(H) : "keyword c" == e ? h(b) : "(" == e ? h(g(")"), b, y(")"), v, w) : "operator" == e ? h(C) : "[" == e ? h(g("]"), T(C, "]"), v, w) : "{" == e ? h(g("}"), T(M, "}"), v, w) : h()
    }

    function b(e) {
      return e.match(/[;\}\)\],]/) ? f() : f(C)
    }

    function w(e, t) {
      if ("operator" == e && /\+\+|--/.test(t)) return h(w);
      if ("operator" == e && "?" == t) return h(C, y(":"), C);
      if (";" != e) return "(" == e ? h(g(")"), T(C, ")"), v, w) : "." == e ? h(k, w) : "[" == e ? h(g("]"), C, y("]"), v, w) : void 0
    }

    function S(e) {
      return ":" == e ? h(v, x) : f(w, y(";"), v)
    }

    function k(e) {
      return "variable" == e ? (Y.marked = "property", h()) : void 0
    }

    function M(e) {
      return "variable" == e && (Y.marked = "property"), U.hasOwnProperty(e) ? h(y(":"), C) : void 0
    }

    function T(e, t) {
      function n(r) {
        return "," == r ? h(e, n) : r == t ? h() : h(y(t))
      }
      return function (r) {
        return r == t ? h() : f(e, n)
      }
    }

    function A(e) {
      return "}" == e ? h() : f(x, A)
    }

    function E(e) {
      return ":" == e ? h(W) : f()
    }

    function W(e) {
      return "variable" == e ? (Y.marked = "variable-3", h()) : f()
    }

    function L(e, t) {
      return "variable" == e ? (d(t), V ? h(E, N) : h(N)) : f()
    }

    function N(e, t) {
      return "=" == t ? h(C, N) : "," == e ? h(L) : void 0
    }

    function D(e) {
      return "var" == e ? h(L, y(";"), O) : ";" == e ? h(O) : "variable" == e ? h(z) : h(O)
    }

    function z(e, t) {
      return "in" == t ? h(C) : h(w, O)
    }

    function O(e, t) {
      return ";" == e ? h(I) : "in" == t ? h(C) : h(C, y(";"), I)
    }

    function I(e) {
      ")" != e && h(C)
    }

    function H(e, t) {
      return "variable" == e ? (d(t), h(H)) : "(" == e ? h(g(")"), p, T(P, ")"), v, x, m) : void 0
    }

    function P(e, t) {
      return "variable" == e ? (d(t), V ? h(E) : h()) : void 0
    }
    var B, R, j = e.indentUnit,
      F = t.json,
      V = t.typescript,
      G = function () {
        function e(e) {
          return {
            type: e,
            style: "keyword"
          }
        }
        var t = e("keyword a"),
          n = e("keyword b"),
          r = e("keyword c"),
          i = e("operator"),
          o = {
            type: "atom",
            style: "atom"
          }, a = {
            "if": t,
            "while": t,
            "with": t,
            "else": n,
            "do": n,
            "try": n,
            "finally": n,
            "return": r,
            "break": r,
            "continue": r,
            "new": r,
            "delete": r,
            "throw": r,
            "var": e("var"),
            "const": e("var"),
            let: e("var"),
            "function": e("function"),
            "catch": e("catch"),
            "for": e("for"),
            "switch": e("switch"),
            "case": e("case"),
            "default": e("default"),
            "in": i,
            "typeof": i,
            "instanceof": i,
            "true": o,
            "false": o,
            "null": o,
            undefined: o,
            NaN: o,
            Infinity: o
          };
        if (V) {
          var l = {
            type: "variable",
            style: "variable-3"
          }, s = {
              "interface": e("interface"),
              "class": e("class"),
              "extends": e("extends"),
              constructor: e("constructor"),
              "public": e("public"),
              "private": e("private"),
              "protected": e("protected"),
              "static": e("static"),
              "super": e("super"),
              string: l,
              number: l,
              bool: l,
              any: l
            };
          for (var u in s) a[u] = s[u]
        }
        return a
      }(),
      K = /[+\-*&%=<>!?|]/,
      U = {
        atom: !0,
        number: !0,
        variable: !0,
        string: !0,
        regexp: !0
      }, Y = {
        state: null,
        column: null,
        marked: null,
        cc: null
      }, $ = {
        name: "this",
        next: {
          name: "arguments"
        }
      };
    return v.lex = !0, {
      startState: function (e) {
        return {
          tokenize: o,
          lastType: null,
          cc: [],
          lexical: new s((e || 0) - j, 0, "block", !1),
          localVars: t.localVars,
          context: t.localVars && {
            vars: t.localVars
          },
          indented: 0
        }
      },
      token: function (e, t) {
        if (e.sol() && (t.lexical.hasOwnProperty("align") || (t.lexical.align = !1), t.indented = e.indentation()), e.eatSpace()) return null;
        var n = t.tokenize(e, t);
        return "comment" == B ? n : (t.lastType = B, c(t, n, B, R, e))
      },
      indent: function (e, t) {
        if (e.tokenize == l) return CodeMirror.Pass;
        if (e.tokenize != o) return 0;
        var n = t && t.charAt(0),
          r = e.lexical;
        "stat" == r.type && "}" == n && (r = r.prev);
        var i = r.type,
          a = n == i;
        return "vardef" == i ? r.indented + ("operator" == e.lastType || "," == e.lastType ? 4 : 0) : "form" == i && "{" == n ? r.indented : "form" == i ? r.indented + j : "stat" == i ? r.indented + ("operator" == e.lastType || "," == e.lastType ? j : 0) : "switch" != r.info || a ? r.align ? r.column + (a ? 0 : 1) : r.indented + (a ? 0 : j) : r.indented + (/^(?:case|default)\b/.test(t) ? j : 2 * j)
      },
      electricChars: ":{}"
    }
  }), CodeMirror.defineMIME("text/javascript", "javascript"), CodeMirror.defineMIME("application/json", {
    name: "javascript",
    json: !0
  }), CodeMirror.defineMIME("text/typescript", {
    name: "javascript",
    typescript: !0
  }), CodeMirror.defineMIME("application/typescript", {
    name: "javascript",
    typescript: !0
  }), CodeMirror.overlayMode = CodeMirror.overlayParser = function (e, t, n) {
  return {
    startState: function () {
      return {
        base: CodeMirror.startState(e),
        overlay: CodeMirror.startState(t),
        basePos: 0,
        baseCur: null,
        overlayPos: 0,
        overlayCur: null
      }
    },
    copyState: function (n) {
      return {
        base: CodeMirror.copyState(e, n.base),
        overlay: CodeMirror.copyState(t, n.overlay),
        basePos: n.basePos,
        baseCur: null,
        overlayPos: n.overlayPos,
        overlayCur: null
      }
    },
    token: function (r, i) {
      return r.start == i.basePos && (i.baseCur = e.token(r, i.base), i.basePos = r.pos), r.start == i.overlayPos && (r.pos = r.start, i.overlayCur = t.token(r, i.overlay), i.overlayPos = r.pos), r.pos = Math.min(i.basePos, i.overlayPos), r.eol() && (i.basePos = i.overlayPos = 0), null == i.overlayCur ? i.baseCur : null != i.baseCur && n ? i.baseCur + " " + i.overlayCur : i.overlayCur
    },
    indent: e.indent && function (t, n) {
      return e.indent(t.base, n)
    },
    electricChars: e.electricChars,
    innerMode: function (t) {
      return {
        state: t.base,
        mode: e
      }
    },
    blankLine: function (n) {
      e.blankLine && e.blankLine(n.base), t.blankLine && t.blankLine(n.overlay)
    }
  }
},

function () {
  "use strict";

  function e(e, t) {
    var n = "" + e,
      r = 2 === arguments.length ? t : 2;
    if (n.length > r) return n;
    for (; r > n.length; n = "0" + n);
    return n
  }

  function t(e, t) {
    var r = 2 === arguments.length && t ? t instanceof Date ? t : new Date(t) : new Date,
      i = new String(e);
    return i.replace(o, function (e) {
        return n[e](r)
      })
  }
  var n = {
    YYYY: function (e) {
      return e.getFullYear()
    },
    YY: function (t) {
      return e(t.getFullYear() % 100)
    },
    MM: function (t) {
      return e(t.getMonth() + 1)
    },
    M: function (e) {
      return e.getMonth() + 1
    },
    dd: function (t) {
      return e(t.getDate())
    },
    d: function (e) {
      return e.getDate()
    },
    hh: function (t) {
      return e(t.getHours())
    },
    h: function (e) {
      return e.getHours()
    },
    mm: function (t) {
      return e(t.getMinutes())
    },
    m: function (e) {
      return e.getMinutes()
    },
    ss: function (t) {
      return e(t.getSeconds())
    },
    s: function (e) {
      return e.getSeconds()
    },
    ff: function (t) {
      return e(t.getMilliseconds(), 3)
    },
    f: function (e) {
      return e.getMilliseconds()
    },
    Z: function (t) {
      var n = t.getTimezoneOffset(),
        r = Math.abs(Math.floor(n / 60)),
        i = n % 60,
        o = n > 0 ? "+" : "-";
      return [o, e(r), ":", e(i)].join("")
    }
  }, r = [];
  for (var i in n) n.hasOwnProperty(i) && r.push(i);
  var o = RegExp(r.join("|"), "mg"),
    a = t.createFormatter = function (e) {
      return function (n) {
        return t(e, n)
      }
    }, l = t.formatters = {}, s = t.register = function (e, t) {
      return l[e] = a(t)
    };
  s("ISODate", "YYYY-MM-dd"), s("ISOTime", "hh:mm:ss"), s("ISODateTime", "YYYY-MM-ddThh:mm:ss"), s("ISODateTimeTZ", "YYYY-MM-ddThh:mm:ss Z");
  var u;
  u = "undefined" != typeof window ? window : "undefined" != typeof global ? global : this;
  var c = u.datef;
  t.noConflict = function () {
    return u.datef = c, this
  }, "undefined" != typeof exports ? module.exports.datef = t : "function" == typeof define && define.amd ? define("datef", [], function () {
      return t
    }) : u.datef = t
}(), "function" != typeof Object.create && (Object.create = function (e) {
    "use strict";

    function t() {}
    return t.prototype = e, new t
  }),

function (e) {
  "use strict";
  var t = {
    init: function (t, n) {
      this.options = e.extend({}, e.fn.resizer.options, t), this.$elem = e(n), this.$handle = this.$elem.find(this.options.handle), this.$container = e(this.options.container || "body"), this.$alsoResize = e(this.options.alsoResize), this.options.containerWidth = this.$container.width(), this.$handle.on("mousedown", e.proxy(this.start, this)).on("mouseup", e.proxy(this.stop, this))
    },
    start: function (t) {
      var n = this.$handle.outerWidth(),
        r = this.$handle.offset().left + n - t.pageX;
      1 >= this.options.maxWidth && (this.maxWidth = Math.ceil(this.$container.width() * this.options.maxWidth)), 1 >= this.options.minWidth && (this.minWidth = Math.ceil(this.$container.width() * this.options.minWidth)), this.$container.addClass("resizing").on("mousemove", e.proxy(this.resize, this, n, r))
    },
    resize: function (e, t, n) {
      var r = n.pageX + t - e,
        i = this.$elem.offset().left,
        o = r - i + e,
        a = this.minWidth || 10,
        l = this.maxWidth && this.options.containerWidth > this.maxWidth ? this.maxWidth : this.options.containerWidth;
      o >= a && l >= o && (this.$handle.offset({
            left: r
          }), this.$elem.width(o), this.$alsoResize.length && this.$alsoResize.width(this.options.containerWidth - o))
    },
    stop: function () {
      this.$container.removeClass("resizing").off("mousemove")
    }
  };
  e.fn.resizer = function (n) {
    return this.each(function () {
        var r = Object.create(t);
        r.init(n, this), e.data(this, "resizer", r)
      })
  }, e.fn.resizer.options = {
    handle: ".handle",
    maxWidth: null,
    minWidth: null,
    glueToSide: !0,
    alsoResize: null,
    container: null
  }
}(jQuery),

function (e, t, n) {
  function r(e, n) {
    var r, i = t.createElement(e || "div");
    for (r in n) i[r] = n[r];
    return i
  }

  function i(e) {
    for (var t = 1, n = arguments.length; n > t; t++) e.appendChild(arguments[t]);
    return e
  }

  function o(e, t, n, r) {
    var i = ["opacity", t, ~~ (100 * e), n, r].join("-"),
      o = .01 + 100 * (n / r),
      a = Math.max(1 - (1 - e) / t * (100 - o), e),
      l = c.substring(0, c.indexOf("Animation")).toLowerCase(),
      s = l && "-" + l + "-" || "";
    return h[i] || (d.insertRule("@" + s + "keyframes " + i + "{" + "0%{opacity:" + a + "}" + o + "%{opacity:" + e + "}" + (o + .01) + "%{opacity:1}" + (o + t) % 100 + "%{opacity:" + e + "}" + "100%{opacity:" + a + "}" + "}", d.cssRules.length), h[i] = 1), i
  }

  function a(e, t) {
    var r, i, o = e.style;
    if (o[t] !== n) return t;
    for (t = t.charAt(0).toUpperCase() + t.slice(1), i = 0; f.length > i; i++) if (r = f[i] + t, o[r] !== n) return r
  }

  function l(e, t) {
    for (var n in t) e.style[a(e, n) || n] = t[n];
    return e
  }

  function s(e) {
    for (var t = 1; arguments.length > t; t++) {
      var r = arguments[t];
      for (var i in r) e[i] === n && (e[i] = r[i])
    }
    return e
  }

  function u(e) {
    for (var t = {
        x: e.offsetLeft,
        y: e.offsetTop
      }; e = e.offsetParent;) t.x += e.offsetLeft, t.y += e.offsetTop;
    return t
  }
  var c, f = ["webkit", "Moz", "ms", "O"],
    h = {}, d = function () {
      var e = r("style", {
          type: "text/css"
        });
      return i(t.getElementsByTagName("head")[0], e), e.sheet || e.styleSheet
    }(),
    p = {
      lines: 12,
      length: 7,
      width: 5,
      radius: 10,
      rotate: 0,
      corners: 1,
      color: "#000",
      speed: 1,
      trail: 100,
      opacity: .25,
      fps: 20,
      zIndex: 2e9,
      className: "spinner",
      top: "auto",
      left: "auto",
      position: "relative"
    }, m = function m(e) {
      return this.spin ? (this.opts = s(e || {}, m.defaults, p), n) : new m(e)
    };
  m.defaults = {}, s(m.prototype, {
      spin: function (e) {
        this.stop();
        var t, n, i = this,
          o = i.opts,
          a = i.el = l(r(0, {
                className: o.className
              }), {
              position: o.position,
              width: 0,
              zIndex: o.zIndex
            }),
          s = o.radius + o.length + o.width;
        if (e && (e.insertBefore(a, e.firstChild || null), n = u(e), t = u(a), l(a, {
                left: ("auto" == o.left ? n.x - t.x + (e.offsetWidth >> 1) : parseInt(o.left, 10) + s) + "px",
                top: ("auto" == o.top ? n.y - t.y + (e.offsetHeight >> 1) : parseInt(o.top, 10) + s) + "px"
              })), a.setAttribute("aria-role", "progressbar"), i.lines(a, i.opts), !c) {
          var f = 0,
            h = o.fps,
            d = h / o.speed,
            p = (1 - o.opacity) / (d * o.trail / 100),
            m = d / o.lines;
          (function g() {
              f++;
              for (var e = o.lines; e; e--) {
                var t = Math.max(1 - (f + e * m) % d * p, o.opacity);
                i.opacity(a, o.lines - e, t, o)
              }
              i.timeout = i.el && setTimeout(g, ~~ (1e3 / h))
            })()
        }
        return i
      },
      stop: function () {
        var e = this.el;
        return e && (clearTimeout(this.timeout), e.parentNode && e.parentNode.removeChild(e), this.el = n), this
      },
      lines: function (e, t) {
        function n(e, n) {
          return l(r(), {
              position: "absolute",
              width: t.length + t.width + "px",
              height: t.width + "px",
              background: e,
              boxShadow: n,
              transformOrigin: "left",
              transform: "rotate(" + ~~(360 / t.lines * s + t.rotate) + "deg) translate(" + t.radius + "px" + ",0)",
              borderRadius: (t.corners * t.width >> 1) + "px"
            })
        }
        for (var a, s = 0; t.lines > s; s++) a = l(r(), {
              position: "absolute",
              top: 1 + ~(t.width / 2) + "px",
              transform: t.hwaccel ? "translate3d(0,0,0)" : "",
              opacity: t.opacity,
              animation: c && o(t.opacity, t.trail, s, t.lines) + " " + 1 / t.speed + "s linear infinite"
            }), t.shadow && i(a, l(n("#000", "0 0 4px #000"), {
                top: "2px"
              })), i(e, i(a, n(t.color, "0 0 1px rgba(0,0,0,.1)")));
        return e
      },
      opacity: function (e, t, n) {
        e.childNodes.length > t && (e.childNodes[t].style.opacity = n)
      }
    }),

  function () {
    function e(e, t) {
      return r("<" + e + ' xmlns="urn:schemas-microsoft.com:vml" class="spin-vml">', t)
    }
    var t = l(r("group"), {
        behavior: "url(#default#VML)"
      });
    !a(t, "transform") && t.adj ? (d.addRule(".spin-vml", "behavior:url(#default#VML)"), m.prototype.lines = function (t, n) {
        function r() {
          return l(e("group", {
                coordsize: u + " " + u,
                coordorigin: -s + " " + -s
              }), {
              width: u,
              height: u
            })
        }

        function o(t, o, a) {
          i(f, i(l(r(), {
                  rotation: 360 / n.lines * t + "deg",
                  left: ~~o
                }), i(l(e("roundrect", {
                      arcsize: n.corners
                    }), {
                    width: s,
                    height: n.width,
                    left: n.radius,
                    top: -n.width >> 1,
                    filter: a
                  }), e("fill", {
                    color: n.color,
                    opacity: n.opacity
                  }), e("stroke", {
                    opacity: 0
                  }))))
        }
        var a, s = n.length + n.width,
          u = 2 * s,
          c = 2 * -(n.width + n.length) + "px",
          f = l(r(), {
              position: "absolute",
              top: c,
              left: c
            });
        if (n.shadow) for (a = 1; n.lines >= a; a++) o(a, -2, "progid:DXImageTransform.Microsoft.Blur(pixelradius=2,makeshadow=1,shadowopacity=.3)");
        for (a = 1; n.lines >= a; a++) o(a);
        return i(t, f)
      }, m.prototype.opacity = function (e, t, n, r) {
        var i = e.firstChild;
        r = r.shadow && r.lines || 0, i && i.childNodes.length > t + r && (i = i.childNodes[t + r], i = i && i.firstChild, i = i && i.firstChild, i && (i.opacity = n))
      }) : c = a(t, "animation")
  }(), "function" == typeof define && define.amd ? define(function () {
      return m
    }) : e.Spinner = m
}(window, document);
var JSONGenerator = function (global, undefined) {
  "use strict";
  Array.isArray || (Array.isArray = function (e) {
      return "[object Array]" === Object.prototype.toString.call(e)
    }), Object.prototype.isEmpty = function () {
    for (var e in this) if (this.hasOwnProperty(e)) return !1;
    return !0
  }, Object.defineProperty(Object.prototype, "isEmpty", {
      enumerable: !1
    }), global.JsonGenerator = {
    init: function (e) {
      return this.config = e, this.parseFieldRegex = /\{\{?(\w+)(?:\((-?\d*(?:\.?)\d+?)?((\w*[\-:\.\s]?)*)?.?\s*?(-?\d*(?:\.?)\d+?)?\))?\}\}/g, this.makeData(), this
    },
    makeData: function () {
      var e = this.generate.cfg.data;
      for (var t in e) e.hasOwnProperty(t) && (e[t] = e[t].split(","))
    },
    makeRandoms: function () {
      var e = {}, t = this.generate.cfg.data;
      for (var n in t) t.hasOwnProperty(n) && (e[n] = this.generate.numeric(0, t[n].length - 1));
      this.generate.cfg.random = e
    },
    generateObjects: function (e, t) {
      for (var n = [], r = 1; t >= r; r++) n.push(this.parseObject(e, r));
      return n
    },
    cloneObject: function (e) {
      if (!e || "object" != typeof e) return e;
      var t, n, r = "function" == typeof e.pop ? [] : {};
      for (t in e) e.hasOwnProperty(t) && (n = e[t], r[t] = n && "object" == typeof n ? this.cloneObject(n) : n);
      return r
    },
    parseObject: function (e, t) {
      e = this.cloneObject(e), this.makeRandoms();
      for (var n in e) e.hasOwnProperty(n) && (e[n] = this.parseFieldValues(e[n], t));
      return e
    },
    parseArray: function (e) {
      if (this.stringHasTags(e[0])) {
        var t = this.parseStringParams(e.shift()).filter(function (e) {
            return e
          }),
          n = e[0],
          r = parseInt(t[2], 10),
          i = parseInt(t[3], 10),
          o = r;
        if (isNaN(i) || (o = this.generate.numeric(r, i)), !n || 0 === n.length || n.isEmpty()) return [];
        if ("object" == typeof n) e = this.generateObjects(n, o);
        else {
          e = [];
          for (var a = 0; o > a; a++) e.push(this.parseFieldValues(n, a))
        }
      }
      return e
    },
    stringHasTags: function (e) {
      return e ? -1 !== e.indexOf("{{") : !1
    },
    parseFieldValues: function (e, t) {
      return Array.isArray(e) ? e = this.parseArray(e) : "object" == typeof e ? e = this.parseObject(e, 0) : "string" == typeof e ? e = this.parseString(e, t) : "function" == typeof e && (e = e.call(this.generate, t)), e
    },
    parseStringParams: function (e) {
      var t = null;
      return this.stringHasTags(e) && (t = (this.parseFieldRegex.exec(e) || []).filter(function () {
            return !0
          }), this.parseFieldRegex.lastIndex = 0), t
    },
    parseString: function (e, t) {
      var n = this;
      if (n.stringHasTags(e)) {
        var r = "";
        e = e.replace(n.parseFieldRegex, function () {
            var e = Array.prototype.slice.call(arguments).filter(function (e) {
                return e
              });
            return r = e[1], n.generate[r](e[2], e[3], t)
          }), "bool" === r ? e = "true" === e : "date" !== r && !isNaN(parseFloat(e)) && isFinite(e) && (e = +e)
      }
      return e
    },
    generate: {
      cfg: {
        data: {
          firstNames: "Isabella,Emma,Olivia,Sophia,Ava,Emily,Madison,Abigail,Chloe,Mia,Elizabeth,Addison,Alexis,Ella,Samantha,Natalie,Grace,Lily,Alyssa,Ashley,Sarah,Taylor,Hannah,Brianna,Hailey,Kaylee,Lillian,Leah,Anna,Allison,Victoria,Avery,Gabriella,Nevaeh,Kayla,Sofia,Brooklyn,Riley,Evelyn,Savannah,Aubrey,Alexa,Peyton,Makayla,Layla,Lauren,Zoe,Sydney,Audrey,Julia,Jasmine,Arianna,Claire,Brooke,Amelia,Morgan,Destiny,Bella,Madelyn,Katherine,Kylie,Maya,Aaliyah,Madeline,Sophie,Kimberly,Kaitlyn,Charlotte,Alexandra,Jocelyn,Maria,Valeria,Andrea,Trinity,Zoey,Gianna,Mackenzie,Jessica,Camila,Faith,Autumn,Ariana,Genesis,Payton,Bailey,Angelina,Caroline,Mariah,Katelyn,Rachel,Vanessa,Molly,Melanie,Serenity,Khloe,Gabrielle,Paige,Mya,Eva,Isabelle",
          lastNames: "Abramson,Hoggarth,Adamson,Brickman,Mercer,Brooks,Michaelson,Brown,Miers,Bush,Miller,Calhoun,Miln,Campbell,Milton,Carey,Molligan,Carrington,Morrison,Carroll,Murphy,Carter,Nash,Chandter,Nathan,Chapman,Neal,Charlson,Nelson,Chesterton,Conors,Ogden,Cook,Oldman,Cramer,Oldridge,Creighton,Oliver,Croftoon,Osborne,Crossman,Oswald,Daniels,Otis,Davidson,Owen,Day,Sheldon,Fisher,Sherlock,Smith,Ford,Stanley,Freeman,Thomson,Fulton,Thorndike,Galbraith,Thornton,Gardner,Timmons,Clapton,Gate,Turner,Gerald,Vance,Gibbs,Vaughan,Gilbert,Wainwright,Gill,Walkman,Gilmore,Wallace,Gilson,Waller,Goldman,Ward,Goodman,Warren,Gustman,Watson,Haig,Wayne,Hailey,Webster,Hamphrey,Wesley,Hancock,White,Hardman,WifKinson,Harrison,Winter,Hawkins,Wood,Higgins,Youmans,Hodges,Young",
          companies: "Systheon,Teknoplexon,Pacwest,SysVenamerica,RoboAerlogix,Venconix,Steganoconiche,Interliant,Aluco,Dynarama,Videobanc,Transtouch,Textiqua,Nanobanc,Vencom,Titanirola,Techtron,Unologic,Navivacs,Teraserv,eEyetanic,Orthosoft,Jamrola,SysUSA,Technogra,Superscope,Titanigraf,Allphysiche,Rapigrafix,Ameritron,Westtomik,Cryptotegrity,Multitiqua,US Omnigraphik,Keytheon,Generola,Compuamerica,Orthomedia,Cryptotemplate,Teratopia,Fibrotouch,Skydata,eSteganoergy,Inridium,Xeicon,OpKeycomm,Mescaridic,Netseco,Safetrust,Entcast,Hypervaco,Enlogia,Syssoft,Openserve,Truetomic,Airdyne,Allnet,Jamconik,InfoAirway,iSkyvaco,Pericenta,Netsystems,Robotemplate,Westmedia,iOptystix,Thermotomic,Conrama,iQualcar,Raylog,iEnland,Turbomart,Mescatron,Infragraph,Robotomic,Westgate,Gigaura,Sontopia,Conotomics,Qualserve,US Infratouch,Infraique,Fibrotopia,Unconix,Anagraph,iMedconik,Ventanium,Aprama,Robocomm,Anaframe,Truegate,Tekcar,Indisco,Idmax,Genland,Polytheon,Quintegrity,Celgra,Fibroserve,Safeagra,Proline",
          cities: "Abilene,Akron,Albuquerque,NewMexico,Alexandria,Allentown,AmarilloAnaheim,Anchorage,Alaska,AnnArbor,Antioch,Arlington,Arvada,Athens,Atlanta,Augusta,Aurora,Austin,Bakersfield,Baltimore,Maryland,BatonRouge,Beaumont,Bellevue,Berkeley,Billings,Birmingham,Boise,IdahoBoston,Boulder,Bridgeport,Brownsville,Buffalo,Burbank,Cambridge,CapeCoral,Carrollton,Cary,CedarRapids,Centennial,Chandler,Charleston,Charlotte,Chattanooga,Chesapeake,Chicago,ChulaVista,Cincinnati,Clarksville,Clearwater,Cleveland,Springs,Columbia,Columbus,Concord,CoralSprings,Corona,CorpusChristi,CostaMesa,Dallas,Daly,Davenport,Dayton,Denton,Denver,DesMoines,Detroit,Downey,Durham,ElMonte,ElPaso,Elgin,Elizabeth,ElkGrove,Erie,Escondido,Eugene,Evansville,Fairfield,Fayetteville,Flint,Fontana,FortCollins,FortLauderdale,FortWayne,FortWorth,Fremont,Fresno,Frisco,Fullerton,Gainesville,GardenGrove,Garland,Gilbert,Glendale,GrandPrairie,GrandRapids,GreenBay,Greensboro,Gresham",
          streets: "Walker Street,East Van Buren Street,Madison Avenue,Park Avenue South at 19th Street,Hudson Street,Harrison Street,Wall Street,Water Street,Tudor City Place, nr 43rd Street,Broadway at 88th Street,Greene Street,Bleecker Street,Elm street,Charles Street,Lafayette Street,Mercer Street,Spring Streets,Prince Streets,Wooster Street,Orchard Street,Stanton Streets,Rivington Streets,Thompson Street,Kenmare Streets,Lafayette Streets,Crosby Streets,Broome Streets,Duane Streets,Washington Street,Horatio Streets,Gansevoort Streets,Grand Street"
        },
        random: null
      },
      randomDataItem: function (e, t) {
        return this.cfg.data[e][t === !0 ? this.numeric(0, this.cfg.data[e].length - 1) : this.cfg.random[e]]
      },
      firstName: function (e) {
        return this.randomDataItem("firstNames", e)
      },
      lastName: function (e) {
        return this.randomDataItem("lastNames", e)
      },
      company: function (e) {
        return this.randomDataItem("companies", e)
      },
      city: function (e) {
        return this.randomDataItem("cities", e)
      },
      street: function (e) {
        return this.randomDataItem("streets", e)
      },
      phone: function () {
        return this.numeric(800, 899) + "-" + this.numeric(400, 600) + "-" + this.numeric(2e3, 4e3)
      },
      email: function () {
        var e = this.firstName(),
          t = this.company();
        return (e + "@" + t + ".com").toLowerCase()
      },
      bool: function () {
        return Math.floor(2 * Math.random()) ? !0 : !1
      },
      index: function () {
        return arguments[arguments.length - 1]
      },
      numeric: function (e, t) {
        var n = 0,
          r = (e + "").split(".").pop().length;
        return "n" === e && (e = 1, t = 100), e = +e, t = +t, n = 0 === e % 1 ? Math.round(e - .5 + Math.random() * (t - e + 1)) : (Math.random() * (t - e) + e).toFixed(r), +n
      },
      date: function (e) {
        var t = new Date(1988, 0, 1),
          n = new Date,
          r = new Date(t.getTime() + Math.random() * (n.getTime() - t.getTime()));
        return e = e || "dd.MM.YY", datef(e, r)
      },
      guid: function () {
        var e = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx";
        return e.replace(/[xy]/g, function (e) {
            var t = 0 | 16 * Math.random(),
              n = "x" === e ? t : 8 | 3 & t;
            return n.toString(16)
          })
      },
      lorem: function (e, t) {
        for (var n = "lorem ipsum dolor sit amet consectetuer adipiscing elit sed diam nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam erat volutpat ut wisi enim ad minim veniam quis nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo consequat duis autem vel eum iriure dolor in hendrerit in vulputate velit esse molestie consequat vel illum dolore eu feugiat nulla facilisis at vero eros et accumsan et iusto odio dignissim qui blandit praesent luptatum zzril delenit augue duis dolore te feugait nulla facilisi", r = n.split(" "), i = isNaN(parseInt(t, 10)) ? parseInt(e, 10) : this.numeric(e, t), o = [], a = "", l = function (e) {
              return e ? e.charAt(0).toUpperCase() + e.slice(1) : ""
            }, s = 0; i > s; s++) o.push(r[this.numeric(0, +r.length - 1)]);
        if (i > 1) {
          for (var u = 0; o.length > u; u++) 0 === u ? o[u] = l(o[u]) : u && 0 === u % 5 ? o[u] += "," : u && 0 === u % 14 && (o[u] += ".", o[u + 1] = l(o[u + 1]));
          a = o.join(" ").replace(/\s?[\.,]?$/, ".")
        } else a = o.join("");
        return a
      }
    },
    generateJson: function (template) {
      var tpl = "";
      try {
        tpl = eval(template)
      } catch (error) {
        if (!(error instanceof SyntaxError && "function" == typeof this.config.onError)) throw error;
        return this.config.onError.call(this, error), undefined
      }
      var objectsArr = this.parseArray(tpl);
      this.generatedJson = {
        id: 1,
        jsonrpc: "2.0",
        total: objectsArr.length,
        result: objectsArr
      }, "function" == typeof this.config.onGenerateComplete && this.config.onGenerateComplete.call(this, this.generatedJson)
    },
    getJson: function () {
      return this.generatedJson
    }
  }
}(this);
(function () {
    for (var e, t = function t() {}, n = ["assert", "clear", "count", "debug", "dir", "dirxml", "error", "exception", "group", "groupCollapsed", "groupEnd", "info", "log", "markTimeline", "profile", "profileEnd", "table", "time", "timeEnd", "timeStamp", "trace", "warn"], r = n.length, i = window.console = window.console || {}; r--;) e = n[r], i[e] || (i[e] = t)
  })(),

function (e) {
  function t(e) {
    var t = encodeURIComponent(e).match(/%[89ABab]/g);
    return e.length + (t ? t.length : 0)
  }

  function n(e) {
    function t(e, t) {
      return Array(t + 1).join(e)
    }
    for (var n = "    ", r = "", i = 0, o = !1, a = null, l = 0, s = e.length; s > l; l++) switch (a = e.charAt(l)) {
      case "{":
      case "[":
        o ? r += a : (r += a + "\n" + t(n, i + 1), i += 1);
        break;
      case "}":
      case "]":
        o ? r += a : (i -= 1, r += "\n" + t(n, i) + a);
        break;
      case ",":
        r += o ? a : ",\n" + t(n, i);
        break;
      case ":":
        r += o ? a : ": ";
        break;
      case " ":
      case "\n":
      case "  ":
        o && (r += a);
        break;
      case '"':
        l > 0 && "\\" !== e.charAt(l - 1) && (o = !o), r += a;
        break;
      default:
        r += a
    }
    return r
  }

  function r(e) {
    "localStorage" in window && null !== window.localStorage && localStorage.setItem("lastTemplate", e)
  }

  function i() {
    var t = e("#default").html();
    return "localStorage" in window && null !== window.localStorage ? localStorage.getItem("lastTemplate") || t : t
  }

  function o(t, n) {
    e("#actions").addClass("hide").find(".get-code").addClass("hide"), e("#actions").find(".save").show(), d.clearHistory(), d.setValue(""), e(".code-size").text("").removeClass("over"), e("a.download, #copy-url").attr("href", "#"), s.setText(""), l.setText(""), n || (localStorage.removeItem("lastTemplate"), h.setValue(i())), t.preventDefault()
  }
  var a = 15e4;
  ZeroClipboard.setMoviePath("/js/libs/ZeroClipboard10.swf");
  var l = new ZeroClipboard.Client,
    s = new ZeroClipboard.Client;
  l.glue("copy-code", "copy-code-container"), l.setHandCursor(!0), s.glue("copy-url", "copy-url-container"), s.setHandCursor(!0);
  var u = new Spinner({
      lines: 14,
      length: 8,
      width: 3,
      radius: 8,
      corners: 1,
      rotate: 0,
      color: "#c1c1c2",
      speed: 1.4,
      trail: 60,
      shadow: !1,
      hwaccel: !1,
      className: "spinner",
      zIndex: 10,
      top: "auto",
      left: 0
    }),
    c = new Spinner({
        lines: 15,
        length: 6,
        width: 2,
        radius: 6,
        corners: 0,
        rotate: 0,
        color: "#c1c1c2",
        speed: 1.4,
        trail: 60,
        shadow: !1,
        hwaccel: !1,
        className: "spinner",
        zIndex: 10,
        top: "auto",
        left: "auto"
      });
  jQuery.extend(jQuery.easing, {
      easeOutQuint: function (e, t, n, r, i) {
        return r * ((t = t / i - 1) * t * t * t * t + 1) + n
      },
      easeInQuint: function (e, t, n, r, i) {
        return r * (t /= i) * t * t * t * t + n
      }
    }), CodeMirror.defineMode("jg-template", function (e, t) {
      var n = {
        token: function (e) {
          var t;
          if (e.match("{{")) {
            for (; null != (t = e.next()) && ("}" != t || "}" != e.next()););
            return e.eat("}"), "jg-template"
          }
          for (; null != e.next() && !e.match("{{", !1););
          return null
        }
      };
      return CodeMirror.overlayMode(CodeMirror.getMode(e, t.backdrop || "javascript"), n)
    });
  var f = {
    matchBrackets: !0,
    smartIndent: !0,
    mode: "jg-template",
    indentUnit: 4,
    lineWrapping: !0,
    value: i()
  }, h = new CodeMirror(document.getElementById("template"), e.extend(f, {
          onChange: function (t) {
            var n = t.getValue(),
              i = e(t.getWrapperElement()).prev("h2.start");
            0 !== n.length ? (r(n), i.hide()) : i.fadeIn(300)
          }
        })),
    d = new CodeMirror(document.getElementById("generated"), e.extend(f, {
          mode: "application/json",
          readOnly: !0,
          value: null,
          onChange: function (t) {
            e(t.getWrapperElement()).prev("h2.start").hide()
          }
        })),
    p = JsonGenerator.init({
        onGenerateComplete: function (r) {
          var i = JSON.stringify(r),
            o = n(i),
            s = t(i);
          d.setValue(o), l.setText(o), e(".code-size").text((s / 1024).toFixed(2) + " kB").addClass(s > a ? "over" : ""), s > a && e(".save").hide(), e("#generated h2.start").hide(), e("#actions").hide().removeClass("hide").fadeIn(500)
        },
        onError: function (e) {
          d.setValue(e.name + ": " + e.message)
        }
      });
  e("#generate").on("click", function (e) {
      var t = h.getValue();
      t && (o(e, 1), p.generateJson(t)), e.preventDefault()
    }), e("#template").resizer({
      handle: ".handle",
      alsoResize: "#generated",
      container: "#editors",
      minWidth: .2,
      maxWidth: .8
    }), e(".show-docs").on("click", function (t) {
      var n = e(this),
        r = n.data("col"),
        i = e("#docs"),
        o = i.find("section"),
        a = !i.is(":visible"),
        l = o.is(":visible"),
        s = o.filter('[id="' + r + '"]').is(":visible");
      l && !s ? (e(".show-docs").removeClass("active"), n.addClass("active"), o.hide().filter('[id="' + r + '"]').show()) : (o.hide().filter('[id="' + r + '"]').show(), i.toggle(a), n.toggleClass("active"), e("#editors").animate({
            width: a ? "70%" : "100%"
          }, 500, "easeInQuint")), t.preventDefault()
    });
  var m = e("#feedback");
  e(".show-feedback").on("click", function (t) {
      var n = e(this),
        r = 59 > m.offset().top,
        i = r ? 59 : -350;
      m.is(":animated") || (m.animate({
            top: i
          }, 300, r ? "easeInQuint" : "easeOutQuint"), n.toggleClass("active", r)), t.preventDefault()
    }), m.find("form").on("submit", function (t) {
      var n = e(this);
      c.spin(n.get(0)), e.post("/feedback", n.serialize()).done(function (t) {
          "sent" === t && (c.stop(), n.prev("h2").fadeOut(function () {
                e(this).text("Message sent. Thanks for your Feedback!").fadeIn()
              }))
        }), t.preventDefault()
    }), e("#actions .save").on("click", function (t) {
      var n = e(this),
        r = p.getJson();
      r && (n.addClass("loading"), u.spin(n[0]), e.ajax({
            url: "/j/save",
            type: "post",
            data: {
              json_source: h.getValue(),
              json_string: JSON.stringify(r)
            }
          }).done(function (t) {
            var r = t.url;
            n.removeClass("loading").slideUp(500), u.stop(), s.setText(r), e(".get-code").hide().removeClass("hide").slideDown(500), e("a.download, #copy-url").attr("href", r)
          })), t.preventDefault(t)
    }), e("#actions .reset").on("click", o), e(".btn-reset").on("click", function (e) {
      o(e, !1)
    })
}(jQuery);
